{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Codon Usage and tRNA Anticodon\n",
    "\n",
    "Using a list of genbank files as input, this script:\n",
    "\n",
    "- [x] Extracts all its CDS's\n",
    "- [x] Calculates codon usage\n",
    "\n",
    "---\n",
    "\n",
    "- [x] Converts each sequence to fasta (or only the concatenated tRNAs)\n",
    "- [x] Uses the trnascan-SE to identify the anticodon\n",
    "\n",
    "---\n",
    "\n",
    "Generates an excel spreadsheet, containing the following fields:\n",
    "\n",
    "- [x] Species\n",
    "\n",
    "- [x] Translation table\n",
    "\n",
    "- [x] Aminoacid (Three-letter code)\n",
    "\n",
    "- [x] Codon for that aminoacid\n",
    "\n",
    "- [ ] Has anticodon in mito tRNAs? YES or EMPTY VALUE\n",
    "\n",
    "- [x] Number of occurrences of that codon in the genes \n",
    "\n",
    "- [x] \\1000 - (number_of_codon_occurences / total number of aa - including start_stop_codons) * 1000\n",
    "\n",
    "- [x] Fraction: codon occurences / sum of all codon occurences for that aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's import everything we need:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqUtils import seq3\n",
    "import subprocess, os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order for this script to work, you have to execute it from within the directory with the genbank files or change the working directory using  **os.chdir(\"path/to/directory/\")**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to add all gb files **(no .gbk allowed!)** to a list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genbanks = glob.glob(\"./gbks/*.gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#genbank = !ls *gb\n",
    "#genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# trnascan\n",
    "There are files for the different genetic tables in /home/gabriel/bioinfo/anaconda3/lib/tRNAscan-SE/gcode/gcode.vertmito.\n",
    "Maybe looking at this will give me some ideas on how to run the trnascan for other tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genbank_to_fasta(genbank):\n",
    "    record = SeqIO.read(genbank, \"genbank\")\n",
    "    species = record.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "    seq = record.seq\n",
    "    fasta = \">{}\\n{}\".format(species, seq)\n",
    "    return(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_029755.1 14049\n",
      "NC_028077.1 14402\n",
      "NC_020322.1 13931\n",
      "NC_025223.1 14513\n",
      "NC_024877.1 14316\n",
      "NC_032402.1 14205\n",
      "NC_010777.1 13991\n",
      "NC_024287.1 14601\n",
      "NC_044695.1 14032\n",
      "NC_059873.1 14149\n",
      "NC_008063.1 14436\n",
      "NC_044081.1 14213\n",
      "XXXXXXXX 14877\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/.local/lib/python3.10/site-packages/Bio/GenBank/__init__.py:1380: BiopythonParserWarning: Expected sequence length 14876, found 14877 (XXXXXXXX).\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_028078.1 14193\n",
      "XXXXXXXX 14925\n",
      "NC_033971.1 14776\n",
      "NC_029756.1 14161\n",
      "NC_005925.1 13874\n",
      "NC_044696.1 14687\n",
      "NC_026863.1 16000\n",
      "NC_042829.1 14625\n",
      "NC_020324.1 14459\n",
      "NC_044099.1 14220\n",
      "NC_042902.1 14683\n",
      "NC_028068.1 14639\n",
      "NC_044101.1 14092\n",
      "NC_020323.1 14197\n",
      "XXXXXXXX 13885\n",
      "NC_044653.1 14074\n",
      "NC_027682.1 14575\n",
      "NC_058921.1 14724\n",
      "NC_027492.1 14563\n",
      "NC_060386.1 14610\n",
      "NC_041120.1 14334\n",
      "NC_053648.1 14431\n",
      "NC_005924.1 14215\n",
      "NC_057201.1 14458\n",
      "NC_040861.1 14727\n",
      "NC_053738.1 13874\n",
      "XXXXXXXX 14496\n",
      "NC_026290.1 14156\n",
      "NC_025557.1 14407\n",
      "NC_046736.1 15078\n",
      "NC_025224.1 14442\n",
      "NC_026123.1 14741\n",
      "NC_025775.1 14414\n",
      "NC_025634.1 14617\n",
      "NC_060328.1 14325\n",
      "NC_024281.1 14063\n",
      "NC_005942.1 14381\n",
      "NC_024878.1 14272\n",
      "NC_040859.1 14941\n",
      "NC_031355.1 14783\n"
     ]
    }
   ],
   "source": [
    "for i in genbanks:\n",
    "    seqio = SeqIO.read(i, \"genbank\")\n",
    "    print(seqio.id, len(seqio))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "We also have to calculate codon usage for all CDSs. In order to do so, we first need to extract all CDS from the genbank file, make sure that all sequences are multiple of 3 (or otherwise adding \"A\" residues from poly-A tail to complete the codon) and concatenate the CDS's."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_cds(genbank):\n",
    "    CDS_list = list()\n",
    "    for record in SeqIO.parse(genbank, \"genbank\"):\n",
    "        accession = record.id\n",
    "        species_name = record.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for FEATURE in record.features:\n",
    "            if FEATURE.type == \"CDS\":\n",
    "                gene = FEATURE.qualifiers.get(\"gene\")[0]\n",
    "                sequence = FEATURE.location.extract(record).seq\n",
    "                CDS_list.append([accession, species_name, gene, sequence])\n",
    "    return(CDS_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__le___', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_per_letter_annotations', '_seq', '_set_per_letter_annotations', '_set_seq', 'annotations', 'dbxrefs', 'description', 'features', 'format', 'id', 'letter_annotations', 'lower', 'name', 'reverse_complement', 'seq', 'translate', 'upper']\n",
      "type: CDS\n",
      "location: [444:1359](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508922']\n",
      "    Key: gene, Value: ['ND2']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp12']\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 2']\n",
      "    Key: protein_id, Value: ['YP_009660716.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MMLFLYLISFLFVLNFDEWIFLWLGLELNMFSFLVLVYKKNSIQVFESCFKYFFVQSLGSGLFLGMLYLGCGDLLLLLLLSFKMGVGPFFMWVPSVANGLGWLECGFLLSFQSILPIYLMYMFSSWILWFISFFSLAVGVIGSFHQLELKKLMVYSSIHHLGWMIVCQSVDGQLWLIYLIVYFIILLSVIKAFSDMGFLNLLEYSNNRFSMLIYMLSMGGMPPTLGFILKWLSFLFFIEYDYFLLFVMMLFSVVMLYIYLRIIYDGILLSLSMDWMLVSAKGYSMKELFHLMGGLLIFFFILLV']\n",
      "\n",
      "type: CDS\n",
      "location: [<1497:3030](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508925']\n",
      "    Key: gene, Value: ['COX1']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp11']\n",
      "    Key: product, Value: ['cytochrome c oxidase subunit I']\n",
      "    Key: protein_id, Value: ['YP_009660717.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['LLRWFYSTNHKDIGTLYLLFGAWSGMMGLALSMLLRVELGQLGSFLGDDHVYNVVVTAHALVMIFFMVMPILIGGFGNWLVPLMLGAPDMAFPRMNNLSFWLLPPSLIMLTLSSFSEMGVGAGWTIYPPLSGGDAHYGISVDMAIFSLHLAGASSIMGAINFITTIFNMRSDGMSFEKVPLFVWSVLVTTVLLLLSLPVLAGAITMLLTDRNFNTSFFDPAGGGDPILFQHLFWFFGHPEVYILILPGFGIISHVVSYCVGKREPFGSIGMIYAMVGIGLMGFVVWAHHMFSVGMDVDTRAYFTAATMIIAVPTGIKVFSWMATFHGAYYEMSSPMLWCLGFVYLFTLGGVTGVILANSSLDVVLHDTYYVVAHFHYVLSMGAVFAIMAGINYWFPLFLGVSFSEVKMKLQFYVMFLGVNLTFFPQHFLGLNGMPRRYSDYPDAFYFWNVVSSLGSILSFLGMMYFVMLLWEGLVGKSDMSEYLVSSSLEWQMFLPPLDHTYSQMVLVFR']\n",
      "\n",
      "type: CDS\n",
      "location: [3033:3682](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508909']\n",
      "    Key: gene, Value: ['COX2']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp10']\n",
      "    Key: note, Value: [\"TAA stop codon is completed by the addition of 3' A residues to the mRNA\"]\n",
      "    Key: product, Value: ['cytochrome c oxidase subunit II']\n",
      "    Key: protein_id, Value: ['YP_009660718.1']\n",
      "    Key: transl_except, Value: ['(pos:3682,aa:TERM)']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MPVWGSLYFQNSNSPVMEYLIMFHDYSMIIVILILLMVGYLLVNSIFASYYSIVNREGQELEMIWTIIPGVILVFIAFPSLRLLYLMEESDMFQLSIKTLGHQWFWSYEYAESNLEEFDSYMNEEGDIRLLEVDEMLLLPYDCGSRVIVSSDDVIHSWTVPSMGVKVDAIPGRLNQLMLYIKRPGMYIGQCSEICGANHSFMPISLKVISMFDFLK']\n",
      "\n",
      "type: CDS\n",
      "location: [3774:3921](+)\n",
      "qualifiers:\n",
      "    Key: gene, Value: ['ATP8']\n",
      "\n",
      "type: CDS\n",
      "location: [3917:4562](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508927']\n",
      "    Key: gene, Value: ['ATP6']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp09']\n",
      "    Key: product, Value: ['ATP synthase F0 subunit 6']\n",
      "    Key: protein_id, Value: ['YP_009660719.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MMVSLFSVFDPVVMGFSFNWLVIILCFYIIKPFYFKMMSYNHTMFCMWMMFFYTYLSGVAGQFNKGIVVMGLGMFMYISMMNIMGLFPFIFSGTAHPIITLSLGMMLWLSCFLMGWINVKNSAAHLTPHGSPLILAPLLVLIESVSHLIRPVTLSIRLAANMMAGHLIVGLISSLSMMFAVMCQSLLLMLELGVCIIQGMVFSILFLLYAVEYY']\n",
      "\n",
      "type: CDS\n",
      "location: [4592:5345](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508911']\n",
      "    Key: gene, Value: ['COX3']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp08']\n",
      "    Key: product, Value: ['cytochrome c oxidase subunit III']\n",
      "    Key: protein_id, Value: ['YP_009660720.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MVSMSPWPLLVGLEIFLVMVGFLGMIKSGLNFLFWVSLFMLILTMGLWWRDVVREASFQGEHTVMVMEGLMYGMILFIMSEVFFFASFFWSFFHSSLSPAVELGMHWPPIGLITFSPFQIPLLNTIILLGSGISVTWAHQKLIVGGQAGLSLGLTCFLGLYFLVLQLFEYFYAGFEMSDSVYGSIFFVATGFHGFHVIVGTVFLFIMWLRDFNYHFSKIHHFGFEAAAWYWHFVDVVWLFLFSVVYWWGA']\n",
      "\n",
      "type: CDS\n",
      "location: [5393:5732](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508928']\n",
      "    Key: gene, Value: ['ND3']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp07']\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 3']\n",
      "    Key: protein_id, Value: ['YP_009660721.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MLFLNFLFIMILVMIMMMLMCILKSLEMKSEGNLSAFECGFECINSARCSFSFRFFLIAILFLVFDAEIVLMLPIPYINYGSIEGMILLLFVWILVLGLFYEMSFGVLKWTN']\n",
      "\n",
      "type: CDS\n",
      "location: [6095:7702](-)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508934']\n",
      "    Key: gene, Value: ['ND5']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp06']\n",
      "    Key: note, Value: [\"TAA stop codon is completed by the addition of 3' A residues to the mRNA\"]\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 5']\n",
      "    Key: protein_id, Value: ['YP_009660722.1']\n",
      "    Key: transl_except, Value: ['(pos:complement(6096..6097),aa:TERM)']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MIMMMFLFTFNLTLKYDLKIFSLSWALNTDPNLNMEISLILDWMSLPFASTIIFISSLIFIFSLSYIPKHNHKHFFLMMFLFVTSMLMLVFTDNMIMLILGWDGLGLTSYILVMFYQNANSTASGNITILSNRIGDMMILMAIALMSMTNNFHFMNNENFTLLTTMMLLIAAVSKSAQFPLSAWLPAAMAAPTPISALVHSSTLVTAGTYILMRIMSYPSSSAYILMLLSSMTAFSAGLAANWEQDMKKIIALSTLSQIALISFATSIGQPKLAFFHMNMHAMFKSTLFLNAGYLIHTSSYQDMRMINSQQKNNPSINVMMIINSLALMGLPFLSGYYSKDAIMENMMMSFTNMLASILMIMTLGLTTSYTMRMMNLIIPYTKKNKPMSLIHHDKTMEMPIMMLTPLSIMAGTMTLWNSNNNMIILSISEKITISSIISSSMLISMMLQFKNTKFIHMGLKAIMLWNLHISSTIPMLKTIKMGNKYQKLDSTWNESNPSLYSFSTLYNSKTVMIHKNTFQTLIMLFLIPSLLIIL']\n",
      "\n",
      "type: CDS\n",
      "location: [7764:9001](-)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508935']\n",
      "    Key: gene, Value: ['ND4']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp05']\n",
      "    Key: note, Value: [\"TAA stop codon is completed by the addition of 3' A residues to the mRNA\"]\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 4']\n",
      "    Key: protein_id, Value: ['YP_009660723.1']\n",
      "    Key: transl_except, Value: ['(pos:complement(7765),aa:TERM)']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MFSNSLNKTVVIMTMMMTWLSLSKWNMKIMMWFMMDNISHIFIMLTLYISILIAFSSYNINKLMLPISSSSLILYFAFMYNDLLLFYTAFEMILIPMFIIITINGKQSERLQAGLFLLLYTITASLPLLISIIFINPFSSFSQMNILNYQFPTPILFIMAFLVKMPMFLTHLWLPKAHVEAPVEGSMILAAILLKLGGYGMIRSLPMMYKKIHMMNNWIISISLMGASMTSMNCIRQKDMKSIIALSSVAHMAMVITSLMSMKMYGMKGAMLMMIGHGLISSALFLMINTLYIQNHTRNIIMMKGSKTKCQFTMMWMFLLLIANISAPPTINIMSEIFIFTSSLQWSMLSMIMILITFMLSSSFSMMMYTSIIQGQNNKIPQFNYLNMKTNMAFMAHITPTVMITFKSEILI']\n",
      "\n",
      "type: CDS\n",
      "location: [8891:9290](-)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508915']\n",
      "    Key: gene, Value: ['ND4L']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp04']\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 4L']\n",
      "    Key: protein_id, Value: ['YP_009660724.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MKVMLLLSTISLLNWRNNFIPMLISLEMLLSSVILIMSSKSISLHYTIPIMLIMMVISSSLSLSYLAAMSRTHKSLMTNFIWMVTYVINPCNNIITNIFKFFKQNCSNYNYNNNLTFPIKMKYKNYNMIHNS']\n",
      "\n",
      "type: CDS\n",
      "location: [9338:9764](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508936']\n",
      "    Key: gene, Value: ['ND6']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp03']\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 6']\n",
      "    Key: protein_id, Value: ['YP_009660725.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MVLILICMFLLLMFSHPMSMLLTFGAMVIIIIVSMYMSLGGFWFSYLLMLIIITGLIVIFSYVVSMMPNEYFGMEFFLLALLLFIAILGFENVNLVSDIVSLKGWVIMYNIYSLFMLCVLLILMLTVTWICDFGSGAVLAE']\n",
      "\n",
      "type: CDS\n",
      "location: [9764:10901](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508917']\n",
      "    Key: gene, Value: ['CYTB']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp02']\n",
      "    Key: product, Value: ['cytochrome b']\n",
      "    Key: protein_id, Value: ['YP_009660726.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MLLKSVRSMNLLKLLNGSFVDLASPSNLSYMWNFGSLLGVFLLFQFITGLFLAIHYSCEVMLSFMSIIHIERDVNWGWLNRALHSNGASIFFFVLYLHVGRGLYYGSYMMKNVWLSGVTLFLVSMAVAFLGYVLPWGQMSYWAATVITNLFSAVPFVGFILVEWIWGGFGVSNPTLTRFFSFHYLFPFFMIMLVMFHLFFLHNKGSGNILGIYLDKDKVVFHWFYVVKDVLGLVLLLLLFVFISLEFPYFFMDVENFVMSDPLVTPIHIQPEWYFLFAYAILRSVPNKIGGVLALLVSIMILYTFPFVLVSRFNSVGFYGVAALCFWVFVNVWVLLTWGGSCVVEEPYILVSQIMSILYFFMMFVFSILKDSEDLLMS']\n",
      "\n",
      "type: CDS\n",
      "location: [10997:11879](-)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: db_xref, Value: ['GeneID:40508938']\n",
      "    Key: gene, Value: ['ND1']\n",
      "    Key: locus_tag, Value: ['FKU06_mgp01']\n",
      "    Key: product, Value: ['NADH dehydrogenase subunit 1']\n",
      "    Key: protein_id, Value: ['YP_009660727.1']\n",
      "    Key: transl_table, Value: ['5']\n",
      "    Key: translation, Value: ['MIMLISIMICVAFYTLLERKILSYIQLRKGPNKNSFMGILQPFSDAIKLFNKSNTLILPSNIMLMYMCPTLAMSLALMIITFLPFNLMSHSDNQYNLLSFILVTSMLVYPIIMIGWASNSKFSNLGSIRSVAQMISYEVNLFLILLSIIMIMHSYNINNPLMMSMTWILPGTTIIFLLWSISCLAEVNRSPFDFAEGESELVSGFNIEYMGALFALIFLAEYTGMIILSIMTAVLFSFNSFILMMSFAMLFLWLRGTLPRFRFDLMMKLNWMIMLPLTLINLMPHMNLMSYHF']\n",
      "\n",
      "Total cds:  13\n"
     ]
    }
   ],
   "source": [
    "#Checking if loxosceles similis has 13 CDS's\n",
    "lox = SeqIO.read('./gbks/loxosceles_similis.gb', \"genbank\")\n",
    "print(dir(lox))\n",
    "cds = 0\n",
    "for i in lox.features:\n",
    "    if i.type == \"CDS\":\n",
    "        cds += 1\n",
    "        print(i)\n",
    "print(\"Total cds: \", cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'concatenate_cds' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [9]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m genbanks:\n\u001b[1;32m      3\u001b[0m     genes \u001b[38;5;241m=\u001b[39m extract_cds(i)\n\u001b[0;32m----> 4\u001b[0m     total \u001b[38;5;241m=\u001b[39m \u001b[43mconcatenate_cds\u001b[49m(fix_truncated(genes))\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28mprint\u001b[39m(genes[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m1\u001b[39m], \u001b[38;5;28mlen\u001b[39m(genes), \u001b[38;5;28mlen\u001b[39m(total))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'concatenate_cds' is not defined"
     ]
    }
   ],
   "source": [
    "#All seqs have all 13 CDS's\n",
    "for i in genbanks:\n",
    "    genes = extract_cds(i)\n",
    "    total = concatenate_cds(fix_truncated(genes))\n",
    "    print(genes[0][1], len(genes), len(total))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_truncated(coding_list):\n",
    "    not_trunc_cds_list = list()\n",
    "    for i in coding_list:\n",
    "        truncated_nucs = len(i[3]) % 3\n",
    "        if truncated_nucs:\n",
    "            missing_nucs = 3 - truncated_nucs\n",
    "            #REWRITE THIS IN ORDER TO JUST ADD THE RESIDUES FROM THE POLY-A TAIL - DONE\n",
    "            #print(\"Truncated stop codon for gene {}\".format(i[2]))\n",
    "            #print(\"Before:{} truncated_nucs\".format(len(i[3]) % 3))\n",
    "            #print(\"Before: {}\".format(i[3]))\n",
    "            i[3] = i[3] + (missing_nucs * 'A')\n",
    "            #print()\n",
    "            #print(\"After: {}\".format(i[3]))\n",
    "            #print(\"After:{} truncated_nucs\".format(len(i[3]) % 3))\n",
    "            not_trunc_cds_list.append(i)\n",
    "        else:\n",
    "            not_trunc_cds_list.append(i)\n",
    "    return(not_trunc_cds_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NEXT STEP:\n",
    "\n",
    "Take a look at the \"CAI.py\" file and use it as a reference to write a function that counts codons for any genetic code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adapted from CAI.py:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from itertools import chain\n",
    "#from Bio.Data import CodonTable\n",
    "#from collections import Counter\n",
    "#from Bio.SeqUtils import seq3\n",
    "\n",
    "# get rid of Biopython warning\n",
    "#import warnings\n",
    "#from Bio import BiopythonWarning\n",
    "\n",
    "#warnings.simplefilter(\"ignore\", BiopythonWarning)\n",
    "\n",
    "\n",
    "def set_ncbi_genetic_codes():\n",
    "\n",
    "\n",
    "    ncbi_genetic_codes = {}\n",
    "    for table, codes in CodonTable.unambiguous_dna_by_id.items():\n",
    "        \n",
    "        # invert the genetic code dictionary to map each amino acid to its codons\n",
    "        # create dictionary of synonymous codons;\n",
    "        # Example: {'Phe': ['TTT', 'TTC'], ..., 'End': ['TAA', 'TAG', 'TGA']}\n",
    "        codons_for_amino_acid = {}\n",
    "        for codon, amino_acid in codes.forward_table.items():\n",
    "            amino_acid = seq3(amino_acid)\n",
    "            codons_for_amino_acid[amino_acid] = codons_for_amino_acid.get(amino_acid, [])\n",
    "            codons_for_amino_acid[amino_acid].append(codon)\n",
    "        #Adding the stop codons\n",
    "        codons_for_amino_acid[\"End\"] = CodonTable.unambiguous_dna_by_id[table].stop_codons\n",
    "        \n",
    "        #create dictionary of synonymous codons for each genetic table;\n",
    "        ncbi_genetic_codes[table] = codons_for_amino_acid\n",
    "        \n",
    "    return(ncbi_genetic_codes)\n",
    "\n",
    "#codon_tables = set_ncbi_genetic_codes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Phe': ['TTT', 'TTC'], 'Leu': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'Ser': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC', 'AGA', 'AGG'], 'Tyr': ['TAT', 'TAC'], 'Cys': ['TGT', 'TGC'], 'Trp': ['TGA', 'TGG'], 'Pro': ['CCT', 'CCC', 'CCA', 'CCG'], 'His': ['CAT', 'CAC'], 'Gln': ['CAA', 'CAG'], 'Arg': ['CGT', 'CGC', 'CGA', 'CGG'], 'Ile': ['ATT', 'ATC'], 'Met': ['ATA', 'ATG'], 'Thr': ['ACT', 'ACC', 'ACA', 'ACG'], 'Asn': ['AAT', 'AAC'], 'Lys': ['AAA', 'AAG'], 'Val': ['GTT', 'GTC', 'GTA', 'GTG'], 'Ala': ['GCT', 'GCC', 'GCA', 'GCG'], 'Asp': ['GAT', 'GAC'], 'Glu': ['GAA', 'GAG'], 'Gly': ['GGT', 'GGC', 'GGA', 'GGG'], 'End': ['TAA', 'TAG']}\n"
     ]
    }
   ],
   "source": [
    "print(set_ncbi_genetic_codes().get(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, all tables have exactly 64 codons, and 20 aminoacids (20 + stop) as expected:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'for table in codon_tables.keys():\\n    codons = 0\\n    for aa in codon_tables[table]:\\n        codons += len(codon_tables[table][aa])\\n    print(\"Table number: {} \\t Number of codons: {} \\t Aminoacids: {}\".format          (table, codons , len(codon_tables[table].keys())))'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''for table in codon_tables.keys():\n",
    "    codons = 0\n",
    "    for aa in codon_tables[table]:\n",
    "        codons += len(codon_tables[table][aa])\n",
    "    print(\"Table number: {} \\t Number of codons: {} \\t Aminoacids: {}\".format\\\n",
    "          (table, codons , len(codon_tables[table].keys())))'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_codons(sequence):\n",
    "    from itertools import product\n",
    "    \n",
    "    #create dictionary with each codon for counting (list comprehension inside dictionary comprehension)\n",
    "    codon_count_dict = {codon: 0 for codon in \n",
    "                  [''.join(i) for i in product(\"ATCG\", repeat=3)]}\n",
    "    \n",
    "    if len(sequence) % 3 != 0:\n",
    "        raise ValueError(\"Input sequence not divisible by three\")\n",
    "    ambiguous = 0\n",
    "    for n, i in enumerate(range(0, len(sequence), 3), 1):\n",
    "        codon = str(sequence[i:i+3].upper())\n",
    "        if codon not in codon_count_dict.keys():\n",
    "            ambiguous += 1\n",
    "            continue\n",
    "            #raise ValueError(\"This sequence probably has ambiguous IUPAC characters. \" \n",
    "            #                 \"Please submit sequences containing only A, T, G or C.\")\n",
    "        codon_count_dict[codon] += 1\n",
    "    if ambiguous:\n",
    "        print(\"This sequence probably has ambiguous IUPAC characters.\",\n",
    "                  f\"{ambiguous} codons with such characters have been ignored\")\n",
    "    return(codon_count_dict)\n",
    "\n",
    "#print(count_codons(concat))\n",
    "\n",
    "#codon_count = count_codons(concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# NEXT STEP:\n",
    "Calculate all metrics associated with codon usage AND map it to its corresponding aminoacids, species, etc.\n",
    "Accomodate this data into a dataframe and plot it into an excel file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def per_thousand(codon_count_dict):\n",
    "    total_codon_number = sum(codon_count_dict.values())\n",
    "    per_thousand = {codon: round((count/total_codon_number)*1000, 2) \n",
    "                    for codon, count in codon_count_dict.items()}\n",
    "    return(per_thousand)    \n",
    "\n",
    "#per_thousand = per_thousand(codon_count)\n",
    "#print(per_thousand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def identify_genetic_code(genbank):\n",
    "#    record = SeqIO.read(genbank, \"genbank\")\n",
    "#    genetic_code = set()\n",
    "#    for i in record.features:\n",
    "#        if i.type == \"CDS\":\n",
    "#            genetic_code.add(i.qualifiers.get(\"transl_table\")[0])\n",
    "#    if len(genetic_code) != 1:\n",
    "#        raise ValueError(\"The {} file has more than one genetic code for its CDS's: {} \" \n",
    "#                             \"Please correct that before submiting\".format(genbank, genetic_code))\n",
    "#    return(int(list(genetic_code)[0]))\n",
    "    \n",
    "#genetic_code = identify_genetic_code(genbank[0]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NC_029755.1\n",
      "Name: NC_029755\n",
      "Description: Neoscona nautica mitochondrion, complete genome\n",
      "Database cross-references: BioProject:PRJNA316287\n",
      "Number of features: 76\n",
      "/molecule_type=DNA\n",
      "/topology=circular\n",
      "/data_file_division=INV\n",
      "/date=12-APR-2016\n",
      "/accessions=['NC_029755']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=mitochondrion Neoscona nautica (brown sailor spider)\n",
      "/organism=Neoscona nautica\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Ecdysozoa', 'Arthropoda', 'Chelicerata', 'Arachnida', 'Araneae', 'Araneomorphae', 'Entelegynae', 'Araneoidea', 'Araneidae', 'Neoscona']\n",
      "/references=[Reference(title='The complete mitochondrial genome of Neoscona nautica', ...), Reference(title='Direct Submission', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence is identical to KR259804.\n",
      "COMPLETENESS: full length.\n",
      "Seq('AATATTGTCAGCTAATAAAGCTAATGAGTTCATACCTCATAAATGGAATTATTA...ATT')\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'identify_genetic_code' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [19]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m genbank \u001b[38;5;129;01min\u001b[39;00m genbanks:\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(SeqIO\u001b[38;5;241m.\u001b[39mread(genbank, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenbank\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[0;32m----> 3\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGenetic_code:\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[43midentify_genetic_code\u001b[49m(genbank))\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'identify_genetic_code' is not defined"
     ]
    }
   ],
   "source": [
    "for genbank in genbanks:\n",
    "    print(SeqIO.read(genbank, \"genbank\"))\n",
    "    print(\"Genetic_code:\", identify_genetic_code(genbank))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(codon_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fraction(codon_count_dict, transl_table):\n",
    "    fraction = dict()\n",
    "    for codon_list in transl_table.values():\n",
    "        codon_occurrences_per_aa = 0\n",
    "        for codon in codon_list:\n",
    "            codon_occurrences_per_aa += codon_count_dict[codon]\n",
    "        for codon in codon_list:\n",
    "            fraction[codon] = round(codon_count_dict[codon]/codon_occurrences_per_aa, 2)\n",
    "    return(fraction)\n",
    "\n",
    "#fraction = fraction(codon_count, codon_tables[genetic_code])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(trna_scan_parse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NEXT:\n",
    "Create main function and put all data into a dataframe.\n",
    "**TIP:** In the main function, put the codon usage functions first, then the anticodon.\n",
    "Structure chosen for the dataframe: list of dictionaries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'excel_dict = [\\n    {\\'species\\': species_name, \\'transl_table\\': translat_table,\\n     \\'aminoacid\\': Three-letter aa, \"Has anticodon in mito?\": Yes or no,\\n    \\'Number of codon occurences\\': number of codons, \\'/1000\\': per_thousand,\\n    \\'Fraction\\': fraction}\\n]'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#template:\n",
    "\n",
    "'''excel_dict = [\n",
    "    {'species': species_name, 'transl_table': translat_table,\n",
    "     'aminoacid': Three-letter aa, \"Has anticodon in mito?\": Yes or no,\n",
    "    'Number of codon occurences': number of codons, '/1000': per_thousand,\n",
    "    'Fraction': fraction}\n",
    "]'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [x] Species\n",
    "\n",
    "- [x] Translation table\n",
    "\n",
    "- [x] Aminoacid (Three-letter code)\n",
    "\n",
    "- [x] Codon for that aminoacid\n",
    "\n",
    "- [ ] Has anticodon in mito tRNAs? YES or EMPTY VALUE\n",
    "\n",
    "- [x] Number of occurrences of that codon in the genes \n",
    "\n",
    "- [x] \\1000 - (number_of_codon_occurences / total number of aa - including start_stop_codons) * 1000\n",
    "\n",
    "- [x] Fraction: codon occurences / sum of all codon occurences for that aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_cds(not_trunc):\n",
    "    concat = ''\n",
    "    for i in not_trunc:\n",
    "        concat += i[3]\n",
    "    return concat\n",
    "\n",
    "def species_name(not_trunc):\n",
    "    species = set()\n",
    "    for i in not_trunc:\n",
    "        species.add(i[1])\n",
    "    if len(species) == 1:\n",
    "        return list(species)[0]\n",
    "    else:\n",
    "        raise ValueError(\"More than one organism for a single genbank file?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "coding_list = extract_cds( './gbks/l_laeta.gb')\n",
    "notrunc = fix_truncated(coding_list)\n",
    "concat = concatenate_cds(notrunc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATTATTATCCCCTCTTTTTTGGTTGTTGTTATTTTGTATTTGGTTAGTTTTTTGTTTGTATTTGGTTTCGATGAGTGGTTTTTTATCTGGTTGGGACTGGAGATGAACATGTTTAGATTTGTGTTGTTGGTTTACCGGCGGTTTAACGCCGCGGCTTTGGAGGGGTGTTTTAGGTATTTTTTTGTTCAGAGTTTAGGGTCTGGTTTATTTTTAGGGGGGGTTTATTTAGGTTTAGGCGAGAGTATGATGCTTTTGGTATTGAGATATAAGATAGGGGTAGGTCCTTTCTATTTCTGGTTTCCTCCTGTAGTTGAAAGATTAGAATGGAGAAGTTGTGGTTTGTTAATAACCTTTCAAAAAGTGTTGCCTTTTTATTTGTTTTATATATTTAGAGGTTGGTTAGTTTGGCTAATGGGAGTAATGAGCTTGTTTATTGGTGTTGTGGGATCATTTAATCAATTGAGTTTAAGGAAGTTGATAGCTTTTTCTTCTATTTATCATTTGGGTTGGATGATGATTTGTCAGTGTGTAGATGGGGGGCTTTGGCTCTATTACTTGGTGGTGTATACTGTTTTAATTATTAGTGTTGTTACAATGTTTTTTAGGATAGGGGAAGGAGATTTGGTCCCTAGAAAAGGAAGCCGTCTGAGATTTATAGTAGGGATTTTCAGTATGGGGGGGGTTCCTCCTATATTGGGATTTGTACTAAAATTTATTAGTTTTGTCTATTTTTTAGAGTATGATTATTTTTTATTGTCTTTTATAATTGTTTTATCAATTGTTATGATATACGTTTATATGCGTTTTATTTATAATGAGTTTTTAGGGGTCGAGGAGGTTGCGTGGGTCGGGGGTTGAAATTTAGGTTTAGGTTTGAGAGAGAGAGTGAATTTATTGGGGGGGGTAATTGTTTTGTGTCTCTTGTGATTATTCTTATAATTACTGCGATGATTTTATTCAACTAATCATAAGGATATCGGAACGTTATATTTGTTGTTTGGTGCTTGGGCTGGTATAATAGGACTTGCTTTAAGAATATTGATTCGTCTGGAGTTAGGTCAATTAGGAAGAATTATAGGAGATGATCACGTTTATAATGTTATTGTGACAGCGCATGCTTTAGTTATGATTTTTTTTATAGTGATACCTATTATGATTGGTGGTTTTGGAAATTGGTTGGTTCCATTAATGTTAGGAGCTCCTGATATAGCTTTTCCTCGAATGAATAATTTAAGTTTTTGGTTGTTGCCTCCTTCTTTGTTAATGTTATTATTATCTTCTCTTTGTGAGAGAGGAGTGGGGGCGGGATGAACTATTTATCCACCTCTGTCTGGGGGGCAAGCTCATTTTGGTGTTTCTGTCGATATAGCCATTTTTTCGTTGCATTTAGCAGGTGCTTCTTCTATTATAGGGGCTATTAATTTTATTACAACAGTGATTAATATACGGAGAGAGGGAATGAGATTTGAGAAGGTTCCTTTATTTGTTTGGTCTGTATTAGTAACTGCTGTTTTATTATTATTGTCGTTGCCTGTTTTGGCGGGAGCTATTACAATATTGTTGACTGACCGGAATTTTAATACTTCTTTTTTTGACCCTGCAGGAGGGGGAGATCCTATTTTATTTCAACATTTATTTTGGTTTTTTGGTCATCCGGAAGTTTATATTCTTATTCTTCCAGGATTTGGTATTATTTCTCATGTAATTAGTTACTCGGTAGGGAGGCGGGAGCCATTTGGTTCTATTGGGATAATTTACGCTATAGTTGGTATTGGGATTATGGGTTTTGTGGTCTGGGCTCATCATATGTTTTCTGTGGGCATGGATGTAGACACCCGGGCCTATTTTACAGCGGCTACAATGGTTATTGCTGTTCCAACAGGAATTAAGGTTTTTAGATGAATAGCAACATTACACGGGGCTTATTTTAAGTTGGATAGACCAATATTATGGTGTTTAGGTTTTGTTTATCTGTTTACTTTGGGGGGAATTACTGGTGTCATTTTGGCTAATTCTTCTTTGGATATTGTTCTTCATGATACTTATTATGTTGTTGCTCATTTTCATTATGTGCTGAGTATAGGGGCTGTTTTTGCTATTATGGCAGGGATTAATTATTGGTTTCCTTTATTTTTTGGTGTTGATTTTAGAGGTATCAAGTTAAAAGTACAGTTTTATATTATATTTATTGGGGTAAATATAACCTTTTTTCCTCAGCATTTTTTGGGGTTAAATGGTATGCCGCGGCGTTATTCTGATTACCCAGATGTCTTTTATTTTTGAAATATAATTTCATCTTTGGGGTCAGTTTTGTCTTTGTTAGGGGTTTTGTTTTTTTTAATGATTTTATGGGAAGGTATAGTTTCTAAAAGAGAGTTTGCTGAGTATTTCGTTTCTTCTTCTATGGAGTGGCAGATAAACTTGCCGCCGATGGAGCATACTTATTCACAGTTAGTTTTAGTTAATCGTTAAATGCCGGTTTGAGGTTACTTGTACTTTCAAAATGGGGCTTCTCCGGTAATGGAGCAGCTGATTTTTTTTCATGATCATACGATAATAGTAATAGTGTTAGTGTTAGTTTTGGTTGGGTATGTTTTATTGAGATCTGTATTTATAAATTATTATAGAATTAAGGTTCGTGAAGGACAAGAGTTGGAGAGAATTTGGACTGTGATACCTGCCTTGATTTTGTTGTTAATTGCTTTTCCTTCTTTGCGTCTTTTGTATTTAATAGAAGAGGCAGATAGATGCGAAGTTGGGGTTAGGGTTACAGGGCATCAGTGGTATTGGTCTTATGAATATAGAAATATAATGAAAGAGGAGTTTGATTCTTATATAGCAGAGGGAGATTTTCGGTTGTTAGAGGTAGATGAATATTTAATTATGCCAAGTGAGGTAAATGTTCGCGCTGTTATTACCGCGGAGGATGTGATTCATTCTTGAACGGTGCCGGCGTTTGGGGTAAAAGCAGATGCTGTTCCGGGGCGGTTGAATCAGGTTTTGTATTTTATTAAGCGAGTAGGAATATATTTTGGGCAGTGTTCGGAGATCTGTGGTGCTAACCATAGTTTTATGCCTATTGCTCTTAAATCTATCAGTGTAGGAGATTTTATTAAGTTATTGTAAATGCCACAATTAAGTCCTATATTTTGGGATGTTTATTTTATAGTGGGATTTTTGGTTTTTGTTTATTTATCGATAATAATTAGTGATTTACGTTTTACTGGAAGGGGGGATTTAAATGTTGTTTTGGTTCCAGTTGAGGTGAAGTGATAGGTGGGTTTGTTTTCTGTTTTTGATCCTGTCGTGTTAGGTTTTTCTTTTAATTGGTTAATTATACTTCTTATTATAGCGAGAAGTATACTGTTGTTTTTTAGGGTTAAAAGTTGAGTAGTTTCAATTTATTACGTGTTTTTGACAAACATGTGTAACAGATTAAAGGGGGTAGCAGGATTAAGATATAAAGGTTTGGTGGTTGTGGGGCTAGGATTATTTATGTTTATTATGTTTATAAATTTAGCAGGATTATTTCCTTACGTTTTTACTGGTTCAGCTCATCCAATGGTTACTTTAAGAATGGGGCTTGTTTTATGAATAACTTGTTTCTTGATGGGGTGATCTTTTAATTTTAAAAATAGAGCCGCTCATCTTACCCCGCAAGGGAGACCTTTAGTGTTGTCTCCTTTTTTAGTACTGATTGAATCTGTAAGTCATTTAATTCGACCGTTTACGTTATCTATTCGTTTGGCCGCCAATATAATAGCTGGTCATTTGATTATTGGTTTGATTTCTGGTTTGAAAAGAATATTCTTTTCTGTAGTTGGACAAAGAATTATGTTGATTTTAGAGGGGGGAGTTTGTGTAATTCAAGGTGTGGTTTTTAGTATTTTATTTCTTCTTTATGCGGTTGAATATTATTAGGTGGATAAGAATTTTCATTTGTTTCATATTGTGAGAATTTCTCCATGGCCTCTAATTGTAGGGGGGGAGGTTTTTTTTATTATATTAGGTATATTAAATTTGTTTAAGGGAGGAAGAAGCGCACTTTTTTTTTGTTCTTTTGTTTTGTTGTGTGTTACAAGAGGATTGTGATGGCGAGACGTAGTTCGGGAAGCTAGATTTCAAGGGGAGCATAGAATTATAGTGGTTCACGGATTAATAGTGGGGATGATTTTGTTTATTTTGTCGGAGGTCTTCTTCTTTTTATCATTTTTTTGAGCTTTTTTTCACTCTAGACTTAGTCCAGCAATTGAGTTGGGTTCTGAGTGGCCGCCTGCAGGAATTATTCCTTTTTCTCCTTTCCAGGTTCCTTTATTAAATACTATTATTTTGTTGAGATCGGGGGTGAGTGTAACTTGGGCTCATCAAGAATTGTTAATAGGGCGGAAAATAGATTTTTCATTGGGTTTGACATGGCTCTTAGGGCTTTATTTTCTTAGATTACAGTTGTTTGAATATTTCACTGCGGGTTTCGAGATTAGAGATTCTGTTTACGGTTCTGTTTTTTTTATAGCAACAGGTTTTCATGGTTTTCACGTAATTGTGGGATCTTTATTTTTATTTGTTATGTGATTGCGGGGGTTAATAAATCATTTTTCTAGAATCCATCACTTTGGGTTTGAAGCAGCTGCTTGGTACTGGCATTTTGTGGATGTTGTTTGATTATTTTTATTTTCTGTTGTTTATTGATGGGGCGCTTAAATTTTATTATTTAGAATAATGCTTAGATGCTTATTAGTAGCATTACTTTTAATTTTGATGACGCTTATTAAGGCATGGGAAAGAAAAGTGGATGGAAATTTGTCTTATTTTGAGTGTGGATTTGACAGAATAAGAGTAGTGCGCCGTCCTTTTTCTTACCGTTTTTTTTTGATTGCTATTTTATTTATTGTTTTTGATGTGGAAATTATTTTGATGTTACCCGTTCCTTACGAGGATGTCAGAGGGGTTAGAGGAATTATTTTTCTGTTGTTTTTATGGATCTTATTGGGGGGGCTGTTGTATGAGATGCAGTACGGAGTTTTGTCATGGGCGAGTTAGATAAAAAGTCTGATAATCACCATTCCAATAATTATATCATTCGCATTCCTGGCTACCAATAACTCCTCAATAAATAGACCAGTAGTCATCTCCTGACATCTTAACCCGGACATAAATATAAACCCCCAAATTTCGCTGATTAGCGACTGAATATCATCTCCCTTTGCAGTAACAATCCTTATAATCTCAAGAACATTATTCATTTATAGCCTCTCTTACATCGAAAAAAAAAACCAAAAGTCCTTCTTCATCATCCTCAGTCTATTCATCGCATCAATAATCATTTTAATCTTCACAGACAACATTATTATACTAATCTTAGGATGAGATGGCTTAGGCCTAACATCCTATCTTTTAGTTATATACTACCAAAACCCTAACTCTCTAGCATCAAGCAACATCACCCTCCTATCTAACCGAATCGGAGACATAACTATTATTACAGCAATTGCTATCATATCATCATCTAACACCTTCTCCTTCCCGTATTCATCTATATTCAACGCCACAACGATAATCCTTCTAATAACCGCGGCAGTCTCAAAAAGAGCCCAAATACCTTTATCTGCTTGACTACCGGCGGCCATAGCTGCGCCTACGCCTATTTCAGCCCTTGTTCACTCTTCCACTCTTGTTACAGCAGGCGCCTTTATTATAATCCGTTTAATCAATTATCCCACTGTAACTAGCTATACATTTATACTAATCGCGTGCACAACAGCGCTCTGCGCCGGACTAGCTGCCAACTGAGAACAAGACATAAAAAAAATCATTGCCCTTTCTACCCTTAGTCAAATCGCAATAATCTCTTTTATTGCTTCAATTGGACAACCCACTACAGCATTTTTCCACATAAATACTCACGCTCTGTTCAAATCCACCTTATTCTTAAGAGCCGGCTACCTCATTCACAAAACATCTTATCAAGATTCACGAGCAATTTCCGCCCAGCCGTCATCAACCCCTGCTATCAATAATTCTATCATCATTACAAGATTAGCTCTAATGGGTTTACCCTTCCTTTCCGGTTTCTATTCCAAAGACGCCATTATAGAAAACATAATAATATCAAACACAAATATGACAATCATAATAAGAATATTTTTGGCCGTAGGACTAACAGCTTCCTATTCAACCCGAATAATAAAAATAATCCTAAACCAAAGACAAAAAACCCAACCCATTAACTCGAGCCACTACTCTAATTTCATAAAATTACCAATCATAATAATTTTACCTTTCTCGATTTCATCGGGAACAATTATAATATGATTATCTAACCCACACGCAACAATCATTATCCCCCCTCTACTTAAAACAATGACACTATCTCTAGTCACACTAGGCATCCTACTAGGCCTATTCATAAACTTTAAATCAAAACAATATACCAGCTTAGGACATAAAGCAATCCCAATCTGAAACCTGCATAACTTATCATCGACTCCTTTTTTAAAATTCACTAACATCGGTAAAATTAGAAACCTGTTAGACAACACATGAAACGAATCCAACCCATCCGTCAATTCATTCTCCTACATAAAAATAAGAAGACACAAAATCCCCCGAAATCTTTTATTAACTGCCTTATTTATATCGATTATACTCCCTATCATCTTCATAAATTAAATGCCAGCAACCCTACTCCCTATATTAATCACACCCTTCTGTAAAAAAAGCATACCCAACCTATTAACAACATTGGCACTAACACTCCTATTTACACTACTGAATCCATGAAAATCATCACTTTATTCTCACTGAACAATTATAGATCCCATCACACTAACCCTAATTAGACTTACTATTTACACAACGATCCTTATCATAATATCCTCCTATTCACCTAAAACTATCTTCCCCACCATTATTTCCTCGACCATTATCCTCATCCTGGCATTTTCACTAAATGATATAATTTTATTTTACATCTCCTTTGAACTTATCCTAATCCCTATGTTTATTTTGATTATATCCAATGGAAAACAACCAGAACGCCTCCAAGCCAGCCTCTTCCTGATGTTATATACAATCACAGCCTCCCTCCCGCTTCTCGCCAGAATCATTTTCTTAAATACCTTCAACAACTTCAATATTATGGCAACATCAAACACACAAATAGAAATTCCCATCCTATTAATTTTAGCTTTCCTTGTAAAAATACCTATATTTTTATCCCATTTATGACTTCCTAAAGCTCACGTCGAGGCCCCTGTAGAAGGATCAATAATCCTTGCGGCTATCCTCCTTAAATTAGGAGGTTACGGAATTATCCGCTTCCTCCCTGCTCTAATAAAAAAGATCAATCTAGTAAACAACTGAATCATCAGAATCAGATTGCTAGGCGCATCCCTTACAAGAATTAACTGCATCCGACAAAAAGACATAAAGTCAATCATTGCACTATCCTCAGTCGCTCACATAGCCATAGTTATTTCCAGCCTTCTAACTATAAAACACTCTGGTTTAATAGGGGCCATAGTCATAATAATCGGGCACGGACTAATCTCCTCAGCATTATTTTTCATAATCAATACCTTCTACCTAAAAAAACACACGCGAAACATCCTCCTATTCAAAGGAACAAAAACCTTTTCCAATAACTTAGCAATATGAACATTTATCTTACTAGCAGCCAACATTTCTGCTCCTCCAACAATTAATATCATAAGAGAGCTATTCATTTTCACGTCCTGCATCCAATGATCAACGTCAGCTATGCTACTACTAATAGTAACATTCATATTAACTTCATCCTTTTCCATATTACTATTCTCAATCTTATCGCATAGAAGCATAGACATTATAAACTTCCAAAGACAACTAGATAATAAATCAAACCTGTGCTTCATAAATCACACCATCCCCTCTGTCCTCCTAATTTTAAAAAGAGAGATATTCCTATAAATAAAAATAATATTAATCATTCCCATTATTAGATTATTAAATTGACGTAAAAATATTATTATAGTCCTCGTAAGTCTAGAAATCCTCCTTACAATAGTTATTATAATCATATTACTCTCACATTTATCCATACACTTCGCGATACCAATCATGTTAACAGCTATAGTTATAGGGGCATCATTAGGACTCTCACTACTCGTGACCACCGCACGGTCAAACAAAAAAATAATATCCTCATTCATTTATACACTAAAATATGCCAGCAACCCTACTCCCTATATTAATCACACCCTTCTGTAAGTGGTGATTTTTTTGTGTGGGCTGTTGTTGGCACTAGTATCTCACCCGATAGGTATAGTTTTGGGATTAGGAGTAGTTGTTTTTTGTGTGTGTTTATTGTTTTATATCGAGTTGGGGGGGTTTTGGTTTAGTTATCTTTTATTATTAATAATGATCACAGGACTTATGGTAATTTTTACTTATATAGCGAGTATAGTTCCTAATGAGCGTTTTGGTTTGGACAGAATATTGGTTTGTGTTTTGTTTGGAGTTATGTTGTTGGGGTATGGGGAAGTGCCGGAATATTGGGAAGGGATTTCTTTGCAAGGGTGAGAGGTGATGTTTTATGTCTATAGCCTATTTATGTTGATGGTTTTATTGGTTGTTATGTTGGCGGTTGTTTGGATTTCTGAGTTTTGTTGAGGGGCGGTTCGTGTATATTAAGTGAGGAAGAGTATTCGTAGGGTTGATTTGGTGAATTTGTTGAATGGCTCTTTTGTAGATTTGCCGTCACCTAAAAGATTGAGATACATATGGAATTTTGGTTCTTTGTTGGGGGTGTTTTTAGCTTTTCAGATTGTTACTGGTTTGTTTTTGGCTATTCATTATAGTTGTGATGTGATGTTAGCTTTTGATAGCATTGTTCATATTGAGCGTGACGTAAATTGGGGGTGAATGGATCGTGTTATTCATTCTAACGGTGCAAGAATCTTCTTTTTATTGTTGTATCTTCATATCGGGCGAGGTTTTTATTATGGTTCTTTTTATTTGAAAAATGTTTGGATGAGAGGAGTAACAATTTTTTTGATTTCTATGGGGACAGCTTTTTTGGGTTATGTCTTACCATGAGGGCAGATGTCTTTTTGGGCTGCTACTGTTATTACTAATTTGTTATCTGCTGTTCCTTATATGGGGGTAATGTTGGTTGAGTGAATTTGAGGAGGTTATGCTGTAGGTAATCCGACTTTGACTCGGTTTTTCTCATTTCATTTTTTGTTTCCTTTTGTGTTGTTGGGTTTGGTTGTTGTGCATTTATTCTTTCTTCATAAGTCTGGTTCTAGAAATCCTATGGGGTTGAAATCGGATACCGATAGGGTTGTTTTTCACAGATTATTTGTGGTCAGGGATGTTGTCGGCTTAGTCTTTGTAATGTTTTTGTTGGTGTTAGTATCTTTTGAATTTCCTTATTTATTTATGGATGTGGAGAATTTTTTGTGTGGGGATCCTTTGGTTACACCCGTTCATATTCAACCGGAATGGTATTTTCTTTTTGCTTATGCCATTTTGCGGAGTGTGCCTAATAAAATAGGTGGGGTTATTGCTTTGTTGTTATCTGTTCTTGTTTTGTATTTGTTTCCTGTGATTATGGAAAGTCGCGTGAGGGGGAATGTTTTGTATTTTGTTAGCAGGGTATTGTTTTGGTGGTTTGTGGCAGTTTGACTGTTGTTAACTTGGGGGGGGTCTTGTGTGGTAGAGGATCCTTATATTTTAGTTAGACAGGTTATAAGAGTCGCTTATTTTGTATTTCTATTTATAATAGGTGCTTTTTATAACATAGAAGATCAGATTTTTGATTAGATCTTAAATATCCTAATCACTATCATCGCCATCTTAATTAGCGTTGCTTTCTATACTCTACTCGAACGAAAAATTCTAGGATACATCCAAATTCGAAAAGGACCCAACAAAACATCTACCATCGGGCTACTACAACCAATTAGAGATGCCCTAAAATTATTCAACAAAAGACTGGTTATTACCACACCATCTAACTTTAATCTCACGCTTCTAGGACCTTCAATAGCCATAATTTTCGCCATCCTTACACTAACATTCTTACCTTATAGCCTTAACCCCCATATCGACAATCACATAAACATCCTGACATTAATCCTAATCTCTAGCTTATCTGTCTACCCATTATTATTAACAGGGTGAGCCACAAATTCTAAATATAGCAACATTGGATCTATCCGAAGTGTCGCCCAAATAATCTCTTATGAAGTTAACCTCTTCATAATTATCCTCTCCATAATCCTAATTTTCCACTCATACAGCCTCATAACCAACTTCACGCCAAACTTGTGATTACTACTAGGGTCAACATGGATCTTCGCACTATGGACAATCAGCGCCTTAGCGGAAGTAAACCGCAGACCCTTTGATTTCGCAGAAGGAGAATCAGAATTAGTATCAGGGTTTAATGTAGAGTACATAAGAGCCCCATTTGCGCTAATCTTCCTGGCAGAATATACTAATATAATCGTAATCAGAATATTAACAGCAATACTATTTTCTTTCAATTCACCTACATTAATAATTACATTCACCCTACTTTTCCTTTGAATCCGAGGCGCTCTCCCTCGTTTCCGATTTGATCTCCTAATAAAATTAAACTGAAAAATCATCCTCCCCTTAATCATTATACTTTTAATTCCTAACATAAACTTAATTATATGATCATACTAG'"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Neoscona_nautica'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_name(notrunc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "teste = SeqIO.read(genbanks[0], format = 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__le___', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_per_letter_annotations', '_seq', '_set_per_letter_annotations', '_set_seq', 'annotations', 'dbxrefs', 'description', 'features', 'format', 'id', 'letter_annotations', 'lower', 'name', 'reverse_complement', 'seq', 'translate', 'upper']\n"
     ]
    }
   ],
   "source": [
    "print(dir(teste))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Neoscona nautica'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "teste.annotations.get('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "count = Counter(teste.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__and__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__iadd__', '__iand__', '__init__', '__init_subclass__', '__ior__', '__isub__', '__iter__', '__le__', '__len__', '__lt__', '__missing__', '__module__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__ror__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__weakref__', '_keep_positive', 'clear', 'copy', 'elements', 'fromkeys', 'get', 'items', 'keys', 'most_common', 'pop', 'popitem', 'setdefault', 'subtract', 'total', 'update', 'values']\n"
     ]
    }
   ],
   "source": [
    "print(dir(count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78.77\n",
      "14049\n"
     ]
    }
   ],
   "source": [
    "AT = (count.get(\"A\") + count.get(\"T\"))*100 / count.total()\n",
    "print(round(AT, 2))\n",
    "print(count.total())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_AT(gb):\n",
    "    seqio = SeqIO.read(gb, format=\"genbank\")\n",
    "    species = seqio.annotations.get('organism')\n",
    "    full_seq = seqio.seq\n",
    "    coding_list = extract_cds(gb)\n",
    "    notrunc = fix_truncated(coding_list)\n",
    "    pcg_concat = concatenate_cds(notrunc)\n",
    "    mito_AT = get_perc_AT(full_seq)\n",
    "    pcg_AT = get_perc_AT(pcg_concat)\n",
    "    return ( species, mito_AT, pcg_AT)\n",
    "    \n",
    "def get_perc_AT(seq):\n",
    "    count = Counter(seq)\n",
    "    AT = ( count.get(\"A\") + count.get(\"T\") ) * 100 / count.total()\n",
    "    AT = round(AT, 2)\n",
    "    return AT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/.local/lib/python3.10/site-packages/Bio/GenBank/__init__.py:1380: BiopythonParserWarning: Expected sequence length 14876, found 14877 (XXXXXXXX).\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "for i in genbanks:\n",
    "    #print(calculate_AT(i))\n",
    "    data.append(calculate_AT(i))\n",
    "AT_df = pandas.DataFrame(data, columns=['Species','AT %', 'AT % (PCGs)'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>AT %</th>\n",
       "      <th>AT % (PCGs)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neoscona nautica</td>\n",
       "      <td>78.77</td>\n",
       "      <td>78.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cyrtarachne nagasakiensis</td>\n",
       "      <td>75.70</td>\n",
       "      <td>74.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Phyxioschema suthepium</td>\n",
       "      <td>67.40</td>\n",
       "      <td>66.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pardosa laura</td>\n",
       "      <td>77.42</td>\n",
       "      <td>76.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Plexippus paykulli</td>\n",
       "      <td>73.49</td>\n",
       "      <td>72.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Araneus angulatus</td>\n",
       "      <td>75.14</td>\n",
       "      <td>74.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hypochilus thorelli</td>\n",
       "      <td>70.33</td>\n",
       "      <td>69.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Telamonia vlijmi</td>\n",
       "      <td>77.30</td>\n",
       "      <td>76.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Argiope perforata</td>\n",
       "      <td>74.19</td>\n",
       "      <td>73.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Atypus karschi</td>\n",
       "      <td>73.65</td>\n",
       "      <td>72.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Trichonephila clavata</td>\n",
       "      <td>76.03</td>\n",
       "      <td>74.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Harpactocrates apennicola</td>\n",
       "      <td>71.83</td>\n",
       "      <td>70.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Phoneutria sp.</td>\n",
       "      <td>73.62</td>\n",
       "      <td>72.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Hypsosinga pygmaea</td>\n",
       "      <td>76.10</td>\n",
       "      <td>75.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Loxosceles laeta</td>\n",
       "      <td>66.28</td>\n",
       "      <td>64.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Agelena silvatica</td>\n",
       "      <td>74.46</td>\n",
       "      <td>73.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Neoscona adianta</td>\n",
       "      <td>74.58</td>\n",
       "      <td>73.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Cyriopagopus schmidti</td>\n",
       "      <td>69.80</td>\n",
       "      <td>69.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Cyclosa japonica</td>\n",
       "      <td>72.96</td>\n",
       "      <td>72.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Argyroneta aquatica</td>\n",
       "      <td>71.90</td>\n",
       "      <td>69.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Epeus alboguttatus</td>\n",
       "      <td>77.59</td>\n",
       "      <td>76.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pholcus phalangioides</td>\n",
       "      <td>65.86</td>\n",
       "      <td>65.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Parachtes romandiolae</td>\n",
       "      <td>71.39</td>\n",
       "      <td>70.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Loxosceles similis</td>\n",
       "      <td>72.78</td>\n",
       "      <td>71.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Tetragnatha nitens</td>\n",
       "      <td>74.27</td>\n",
       "      <td>73.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Neoscona scylla</td>\n",
       "      <td>74.63</td>\n",
       "      <td>73.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Liphistius erawan</td>\n",
       "      <td>67.72</td>\n",
       "      <td>66.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Latrodectus curacaviensis</td>\n",
       "      <td>74.04</td>\n",
       "      <td>72.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Neoscona multiplicans</td>\n",
       "      <td>74.82</td>\n",
       "      <td>74.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cyclosa argenteoalba</td>\n",
       "      <td>73.71</td>\n",
       "      <td>73.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Phoneutria boliviensis</td>\n",
       "      <td>71.60</td>\n",
       "      <td>70.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Carrhotus xanthogramma</td>\n",
       "      <td>75.05</td>\n",
       "      <td>74.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Desis jiaxiangi</td>\n",
       "      <td>77.04</td>\n",
       "      <td>75.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Cheliceroides longipalpis</td>\n",
       "      <td>79.04</td>\n",
       "      <td>78.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Oxyopes licenti</td>\n",
       "      <td>78.07</td>\n",
       "      <td>77.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Songthela hangzhouensis</td>\n",
       "      <td>72.22</td>\n",
       "      <td>71.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Phanuelus gladstone</td>\n",
       "      <td>75.09</td>\n",
       "      <td>74.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Mesabolivar sp. ITV1036I3</td>\n",
       "      <td>70.66</td>\n",
       "      <td>69.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Cyriopagopus hainanus</td>\n",
       "      <td>69.65</td>\n",
       "      <td>69.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Latrodectus geometricus</td>\n",
       "      <td>74.42</td>\n",
       "      <td>73.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Neoscona theisi</td>\n",
       "      <td>75.16</td>\n",
       "      <td>74.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Oxytate striatipes</td>\n",
       "      <td>78.21</td>\n",
       "      <td>77.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Oxyopes hupingensis</td>\n",
       "      <td>77.87</td>\n",
       "      <td>77.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Oxyopes sertatus</td>\n",
       "      <td>75.94</td>\n",
       "      <td>75.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Wadicosa fidelis</td>\n",
       "      <td>76.03</td>\n",
       "      <td>75.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Tetragnatha maxillosa</td>\n",
       "      <td>74.50</td>\n",
       "      <td>73.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Araneus ventricosus</td>\n",
       "      <td>73.38</td>\n",
       "      <td>72.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Phintella cavaleriei</td>\n",
       "      <td>78.13</td>\n",
       "      <td>77.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Argiope bruennichi</td>\n",
       "      <td>73.43</td>\n",
       "      <td>72.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Habronattus oregonensis</td>\n",
       "      <td>74.33</td>\n",
       "      <td>73.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Selenops bursarius</td>\n",
       "      <td>74.40</td>\n",
       "      <td>73.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Mesabolivar sp. ITV1036I1</td>\n",
       "      <td>70.59</td>\n",
       "      <td>69.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Dolomedes angustivirgatus</td>\n",
       "      <td>76.78</td>\n",
       "      <td>76.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Species   AT %  AT % (PCGs)\n",
       "0            Neoscona nautica  78.77        78.53\n",
       "1   Cyrtarachne nagasakiensis  75.70        74.98\n",
       "2      Phyxioschema suthepium  67.40        66.61\n",
       "3               Pardosa laura  77.42        76.88\n",
       "4          Plexippus paykulli  73.49        72.79\n",
       "5           Araneus angulatus  75.14        74.42\n",
       "6         Hypochilus thorelli  70.33        69.03\n",
       "7            Telamonia vlijmi  77.30        76.76\n",
       "8           Argiope perforata  74.19        73.34\n",
       "9              Atypus karschi  73.65        72.76\n",
       "10      Trichonephila clavata  76.03        74.93\n",
       "11  Harpactocrates apennicola  71.83        70.97\n",
       "12             Phoneutria sp.  73.62        72.53\n",
       "13         Hypsosinga pygmaea  76.10        75.48\n",
       "14           Loxosceles laeta  66.28        64.42\n",
       "15          Agelena silvatica  74.46        73.55\n",
       "16           Neoscona adianta  74.58        73.93\n",
       "17      Cyriopagopus schmidti  69.80        69.80\n",
       "18           Cyclosa japonica  72.96        72.61\n",
       "19        Argyroneta aquatica  71.90        69.82\n",
       "20         Epeus alboguttatus  77.59        76.91\n",
       "21      Pholcus phalangioides  65.86        65.38\n",
       "22      Parachtes romandiolae  71.39        70.31\n",
       "23         Loxosceles similis  72.78        71.93\n",
       "24         Tetragnatha nitens  74.27        73.07\n",
       "25            Neoscona scylla  74.63        73.73\n",
       "26          Liphistius erawan  67.72        66.76\n",
       "27  Latrodectus curacaviensis  74.04        72.94\n",
       "28      Neoscona multiplicans  74.82        74.06\n",
       "29       Cyclosa argenteoalba  73.71        73.60\n",
       "30     Phoneutria boliviensis  71.60        70.39\n",
       "31     Carrhotus xanthogramma  75.05        74.09\n",
       "32            Desis jiaxiangi  77.04        75.88\n",
       "33  Cheliceroides longipalpis  79.04        78.44\n",
       "34            Oxyopes licenti  78.07        77.67\n",
       "35    Songthela hangzhouensis  72.22        71.50\n",
       "36        Phanuelus gladstone  75.09        74.23\n",
       "37  Mesabolivar sp. ITV1036I3  70.66        69.33\n",
       "38      Cyriopagopus hainanus  69.65        69.63\n",
       "39    Latrodectus geometricus  74.42        73.04\n",
       "40            Neoscona theisi  75.16        74.35\n",
       "41         Oxytate striatipes  78.21        77.72\n",
       "42        Oxyopes hupingensis  77.87        77.46\n",
       "43           Oxyopes sertatus  75.94        75.01\n",
       "44           Wadicosa fidelis  76.03        75.76\n",
       "45      Tetragnatha maxillosa  74.50        73.48\n",
       "46        Araneus ventricosus  73.38        72.63\n",
       "47       Phintella cavaleriei  78.13        77.66\n",
       "48         Argiope bruennichi  73.43        72.79\n",
       "49    Habronattus oregonensis  74.33        73.81\n",
       "50         Selenops bursarius  74.40        73.84\n",
       "51  Mesabolivar sp. ITV1036I1  70.59        69.33\n",
       "52  Dolomedes angustivirgatus  76.78        76.16"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AT_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "AT_df.to_csv(\"AT_percentage.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def invert_codon_table(codon_table):\n",
    "    new_codon_table = dict()\n",
    "    for key in codon_table.keys():\n",
    "        #print(key)\n",
    "        for codon in codon_table.get(key):\n",
    "            new_codon_table[codon] = key\n",
    "            \n",
    "    return(new_codon_table)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def has_anticodon(codon, aa_3letter, anticodon_dict):\n",
    "#    #print(aa_3letter)\n",
    "#    if aa_3letter in anticodon_dict.keys():\n",
    "#        for value in anticodon_dict.get(aa_3letter):\n",
    "#            anticodon = Seq(value, generic_dna)\n",
    "#            if codon == str(anticodon.reverse_complement()):\n",
    "#                return(value)\n",
    "#    return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dataframe_parser(species, genetic_code, inverse_codon_table, \n",
    "                     codon_count_dict, per_thousand_dict, fraction_dict, \n",
    "                     anticodon_dict):\n",
    "    #fields = ['Species', 'Translation Table', 'Aminoacid', 'Codon', 'Anticodon', \n",
    "    #              'Number of Codon Occurences', '/1000', 'Fraction']\n",
    "    fields = ['Species', 'Translation Table', 'Aminoacid', 'Codon', 'Anticodon',\n",
    "                  'Number of Codon Occurences', '/1000', 'Fraction']\n",
    "    dataframe_dict = {k : list() for k in fields}\n",
    "    for codon, fraction in fraction_dict.items():\n",
    "        dataframe_dict.get('Species').append(species)\n",
    "        dataframe_dict.get('Translation Table').append(genetic_code)\n",
    "        aa_3letter = inverse_codon_table.get(codon)\n",
    "        dataframe_dict.get('Aminoacid').append(aa_3letter)\n",
    "        dataframe_dict.get('Codon').append(codon)\n",
    "        dataframe_dict.get('Anticodon').append(has_anticodon(codon, aa_3letter, anticodon_dict))\n",
    "        dataframe_dict.get('Number of Codon Occurences').append(codon_count_dict.get(codon))\n",
    "        dataframe_dict.get('/1000').append(per_thousand_dict.get(codon))\n",
    "        dataframe_dict.get('Fraction').append(fraction)\n",
    "    return(dataframe_dict)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_dataframe(parsed_data):\n",
    "    codon_usage = pandas.DataFrame(parsed_data)\n",
    "    return(codon_usage)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_dataframes(dataframe_data):\n",
    "    dataframes = []\n",
    "    for i in dataframe_data.keys():\n",
    "        dataframes.append(dataframe_data.get(i))\n",
    "    final_dataframe = pandas.concat(dataframes)\n",
    "    return(final_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_dataframe_to_excel(final_dataframe):\n",
    "    final_dataframe.to_excel(\"Codon_usage.xlsx\", index=False, sheet_name=\"codon_usage\")\n",
    "    \n",
    "def export_dataframe_to_csv(final_dataframe):\n",
    "    final_dataframe.to_csv(\"Codon_usage.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "default_anticodons = {'Ala': ['TGC'],\n",
    "                      'Arg': ['TCG'],\n",
    "                      'Asn': ['GTT'],\n",
    "                      'Asp': ['GTC'],\n",
    "                      'Cys': ['GCA'],\n",
    "                      'Gln': ['TTG'],\n",
    "                      'Glu': ['TTC'],\n",
    "                      'Gly': ['TCC'],\n",
    "                      'His': ['GTG'],\n",
    "                      'Ile': ['GAT'],\n",
    "                      'Leu': ['TAA', 'TAG'],\n",
    "                      'Lys': ['CTT'],\n",
    "                      'Met': ['CAT'],\n",
    "                      'Phe': ['GAA'],\n",
    "                      'Pro': ['TGG'],\n",
    "                      'Ser': ['GCT', 'TGA'],\n",
    "                      'Thr': ['TGT'],\n",
    "                      'Trp': ['TCA'],\n",
    "                      'Tyr': ['GTA'],\n",
    "                      'Val': ['TAC']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def trnascan_se_parser(trnascan_output, sequence_name):\n",
    "#    aminoacid = list()\n",
    "#    anticodon = list()\n",
    "#    for i in trnascan_output:\n",
    "#         if i.startswith(sequence_name):\n",
    "#                i = i.split(\"\\t\")\n",
    "#                aminoacid.append(i[4].strip())\n",
    "#                anticodon.append(i[5].strip())\n",
    "#    trnascan_dict = {sequence_name: {i[0]: i[1] for i in zip(aminoacid, anticodon)}}\n",
    "#    return(trnascan_dict)\n",
    "\n",
    "\n",
    "def has_anticodon(codon, aa_3letter, anticodon_dict):\n",
    "    #print(aa_3letter)\n",
    "    if aa_3letter in anticodon_dict.keys():\n",
    "        for value in anticodon_dict.get(aa_3letter):\n",
    "            anticodon = Seq(value)\n",
    "            if codon == str(anticodon.reverse_complement()):\n",
    "                return(value)\n",
    "    return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running analysis for ./gbks/Neoscona_nautica_NC_029755.1.gb\n",
      "Running analysis for ./gbks/Cyrtarachne_nagasakiensis_NC_028077.1.gb\n",
      "Running analysis for ./gbks/Phyxioschema_suthepium_NC_020322.1.gb\n",
      "Running analysis for ./gbks/Pardosa_laura_NC_025223.1.gb\n",
      "Running analysis for ./gbks/Plexippus_paykulli_NC_024877.1.gb\n",
      "Running analysis for ./gbks/Araneus_angulatus_NC_032402.1.gb\n",
      "Running analysis for ./gbks/Hypochilus_thorelli_NC_010777.1.gb\n",
      "This sequence probably has ambiguous IUPAC characters. 1 codons with such characters have been ignored\n",
      "Running analysis for ./gbks/Telamonia_vlijmi_NC_024287.1.gb\n",
      "Running analysis for ./gbks/Argiope_perforata_NC_044695.1.gb\n",
      "Running analysis for ./gbks/Atypus_karschi_NC_059873.1.gb\n",
      "Running analysis for ./gbks/Trichonephila_clavata_NC_008063.1.gb\n",
      "Running analysis for ./gbks/Harpactocrates_apennicola_NC_044081.1.gb\n",
      "Running analysis for ./gbks/phoneutria_sp.gb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/.local/lib/python3.10/site-packages/Bio/GenBank/__init__.py:1380: BiopythonParserWarning: Expected sequence length 14876, found 14877 (XXXXXXXX).\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This sequence probably has ambiguous IUPAC characters. 1 codons with such characters have been ignored\n",
      "Running analysis for ./gbks/Hypsosinga_pygmaea_NC_028078.1.gb\n",
      "Running analysis for ./gbks/l_laeta.gb\n",
      "Running analysis for ./gbks/Agelena_silvatica_NC_033971.1.gb\n",
      "Running analysis for ./gbks/Neoscona_adianta_NC_029756.1.gb\n",
      "Running analysis for ./gbks/Cyriopagopus_schmidti_NC_005925.1.gb\n",
      "Running analysis for ./gbks/Cyclosa_japonica_NC_044696.1.gb\n",
      "Running analysis for ./gbks/Argyroneta_aquatica_NC_026863.1.gb\n",
      "This sequence probably has ambiguous IUPAC characters. 4 codons with such characters have been ignored\n",
      "Running analysis for ./gbks/Epeus_alboguttatus_NC_042829.1.gb\n",
      "Running analysis for ./gbks/Pholcus_phalangioides_NC_020324.1.gb\n",
      "Running analysis for ./gbks/Parachtes_romandiolae_NC_044099.1.gb\n",
      "Running analysis for ./gbks/loxosceles_similis.gb\n",
      "Running analysis for ./gbks/Tetragnatha_nitens_NC_028068.1.gb\n",
      "Running analysis for ./gbks/Neoscona_scylla_NC_044101.1.gb\n",
      "Running analysis for ./gbks/Liphistius_erawan_NC_020323.1.gb\n",
      "Running analysis for ./gbks/l_curacaviensis.gb\n",
      "Running analysis for ./gbks/Neoscona_multiplicans_NC_044653.1.gb\n",
      "Running analysis for ./gbks/Cyclosa_argenteoalba_NC_027682.1.gb\n",
      "Running analysis for ./gbks/Phoneutria_boliviensis_NC_058921.1.gb\n",
      "This sequence probably has ambiguous IUPAC characters. 1 codons with such characters have been ignored\n",
      "Running analysis for ./gbks/Carrhotus_xanthogramma_NC_027492.1.gb\n",
      "Running analysis for ./gbks/Desis_jiaxiangi_NC_060386.1.gb\n",
      "Running analysis for ./gbks/Cheliceroides_longipalpis_NC_041120.1.gb\n",
      "Running analysis for ./gbks/Oxyopes_licenti_NC_053648.1.gb\n",
      "Running analysis for ./gbks/Songthela_hangzhouensis_NC_005924.1.gb\n",
      "Running analysis for ./gbks/Phanuelus_gladstone_NC_057201.1.gb\n",
      "Running analysis for ./gbks/Mesabolivar_sp._ITV1036I3_NC_040861.1.gb\n",
      "Running analysis for ./gbks/Cyriopagopus_hainanus_NC_053738.1.gb\n",
      "Running analysis for ./gbks/l_geometricus.gb\n",
      "Running analysis for ./gbks/Neoscona_theisi_NC_026290.1.gb\n",
      "Running analysis for ./gbks/Oxytate_striatipes_NC_025557.1.gb\n",
      "Running analysis for ./gbks/Oxyopes_hupingensis_NC_046736.1.gb\n",
      "Running analysis for ./gbks/Oxyopes_sertatus_NC_025224.1.gb\n",
      "Running analysis for ./gbks/Wadicosa_fidelis_NC_026123.1.gb\n",
      "Running analysis for ./gbks/Tetragnatha_maxillosa_NC_025775.1.gb\n",
      "Running analysis for ./gbks/Araneus_ventricosus_NC_025634.1.gb\n",
      "Running analysis for ./gbks/Phintella_cavaleriei_NC_060328.1.gb\n",
      "Running analysis for ./gbks/Argiope_bruennichi_NC_024281.1.gb\n",
      "Running analysis for ./gbks/Habronattus_oregonensis_NC_005942.1.gb\n",
      "This sequence probably has ambiguous IUPAC characters. 8 codons with such characters have been ignored\n",
      "Running analysis for ./gbks/Selenops_bursarius_NC_024878.1.gb\n",
      "Running analysis for ./gbks/Mesabolivar_sp._ITV1036I1_NC_040859.1.gb\n",
      "Running analysis for ./gbks/Dolomedes_angustivirgatus_NC_031355.1.gb\n",
      "                      Species  Translation Table Aminoacid Codon Anticodon  \\\n",
      "0            Neoscona_nautica                  5       Phe   TTT      None   \n",
      "1            Neoscona_nautica                  5       Phe   TTC       GAA   \n",
      "2            Neoscona_nautica                  5       Leu   TTA       TAA   \n",
      "3            Neoscona_nautica                  5       Leu   TTG      None   \n",
      "4            Neoscona_nautica                  5       Leu   CTT      None   \n",
      "..                        ...                ...       ...   ...       ...   \n",
      "59  Dolomedes_angustivirgatus                  5       Gly   GGC      None   \n",
      "60  Dolomedes_angustivirgatus                  5       Gly   GGA       TCC   \n",
      "61  Dolomedes_angustivirgatus                  5       Gly   GGG      None   \n",
      "62  Dolomedes_angustivirgatus                  5       End   TAA      None   \n",
      "63  Dolomedes_angustivirgatus                  5       End   TAG      None   \n",
      "\n",
      "    Number of Codon Occurences   /1000  Fraction  \n",
      "0                          323   90.20      0.95  \n",
      "1                           18    5.03      0.05  \n",
      "2                          419  117.01      0.87  \n",
      "3                           21    5.86      0.04  \n",
      "4                           32    8.94      0.07  \n",
      "..                         ...     ...       ...  \n",
      "59                           3    0.83      0.02  \n",
      "60                          94   26.12      0.50  \n",
      "61                          31    8.61      0.17  \n",
      "62                          10    2.78      0.77  \n",
      "63                           3    0.83      0.23  \n",
      "\n",
      "[3392 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "#def dataframe_parser(species, genetic_code, inverse_codon_table, \n",
    "#                     codon_count_dict, per_thousand_dict, fraction_dict, \n",
    "#                     anticodon_dict):\n",
    "    \n",
    "def main_func(genbank_list):\n",
    "    codon_tables = set_ncbi_genetic_codes()\n",
    "    dataframe_data = dict()\n",
    "    for index, genbank in enumerate(genbanks, 1):\n",
    "        print(\"Running analysis for {}\".format(genbank))\n",
    "        try:\n",
    "            not_trunc = fix_truncated(extract_cds(genbank))\n",
    "        \n",
    "            cds_concat = concatenate_cds(not_trunc)\n",
    "        \n",
    "            species = species_name(not_trunc)\n",
    "        \n",
    "            codon_count_dict = count_codons(cds_concat)\n",
    "        \n",
    "            per_thousand_dict = per_thousand(codon_count_dict)\n",
    "        \n",
    "            #genetic_code = identify_genetic_code(genbank)\n",
    "            genetic_code = 5\n",
    "                \n",
    "            fraction_dict = fraction(codon_count_dict, codon_tables[genetic_code])\n",
    "        \n",
    "        \n",
    "            #anticodon_dict = trnascan_se_anticodon_parser(trnascan_output, species)\n",
    "            anticodon_dict = default_anticodons\n",
    "        \n",
    "            inverse_codon_table = invert_codon_table(codon_tables[genetic_code])\n",
    "        \n",
    "            parsed_data = dataframe_parser(species, genetic_code, inverse_codon_table, \n",
    "                                           codon_count_dict, per_thousand_dict, fraction_dict, \n",
    "                                           anticodon_dict)\n",
    "        except Exception as e:\n",
    "            print(\"{} analysis failed\".format(genbank))\n",
    "            print(e)\n",
    "            continue\n",
    "        \n",
    "        dataframe_data[index] = create_dataframe(parsed_data)\n",
    "        \n",
    "    final_dataframe = concat_dataframes(dataframe_data)\n",
    "    print(final_dataframe)\n",
    "    \n",
    "    #export_dataframe_to_excel(final_dataframe)\n",
    "    export_dataframe_to_csv(final_dataframe)\n",
    "        \n",
    "        #print(anticodon_dict.keys())\n",
    "        #print(len(anticodon_dict.keys()))\n",
    "        #print(set(inverse_codon_table.values()))\n",
    "        #print(len(set(inverse_codon_table.values())))\n",
    "        #print(species, type(species))\n",
    "        #print(not_trunc)\n",
    "        #print(cds_concat)\n",
    "        #print(\"per_thousand_dict:\\n{}\".format(per_thousand_dict))\n",
    "        #print(\"codon_count_dict:\\n{}\".format(codon_count_dict))\n",
    "        #print(\"genetic_code {}:\\n{}\".format(genetic_code, codon_tables[genetic_code]))    \n",
    "        #print(\"fraction_dict:\\n{}\".format(fraction_dict))\n",
    "'''        fields = ['species', 'transl_table', 'aminoacid', \"Anticodon\", \n",
    "                  'Number of codon occurences', '/1000', 'Fraction']\n",
    "        excel_output_dict = {k : list() for k in fields}\n",
    "'''\n",
    "main_func(genbank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "gb = SeqIO.read(\"./gbks/Oxyopes_sertatus_NC_025224.1.gb\", \"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__add__', '__bool__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__le___', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__radd__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_per_letter_annotations', '_seq', '_set_per_letter_annotations', '_set_seq', 'annotations', 'dbxrefs', 'description', 'features', 'format', 'id', 'letter_annotations', 'lower', 'name', 'reverse_complement', 'seq', 'translate', 'upper']\n"
     ]
    }
   ],
   "source": [
    "print(dir(gb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NC_025224              14442 bp    DNA     circular INV 16-OCT-2014\n",
      "DEFINITION  Oxyopes sertatus voucher 2013-phc-nj-ox-se mitochondrion, complete\n",
      "            genome.\n",
      "ACCESSION   NC_025224\n",
      "VERSION     NC_025224.1\n",
      "DBLINK      BioProject: PRJNA262796\n",
      "KEYWORDS    RefSeq.\n",
      "SOURCE      mitochondrion Oxyopes sertatus\n",
      "  ORGANISM  Oxyopes sertatus\n",
      "            Eukaryota; Metazoa; Ecdysozoa; Arthropoda; Chelicerata; Arachnida;\n",
      "            Araneae; Araneomorphae; Entelegynae; Lycosoidea; Oxyopidae; Oxyopes.\n",
      "REFERENCE   1  (bases 1 to 14442)\n",
      "  AUTHORS   Pan,W.J., Fang,H.Y., Zhang,P. and Pan,H.C.\n",
      "  TITLE     The complete mitochondrial genome of striped lynx spider Oxyopes\n",
      "            sertatus (Araneae: Oxyopidae)\n",
      "  JOURNAL   Mitochondrial DNA, 1-2 (2014) In press\n",
      "   PUBMED   25208169\n",
      "  REMARK    Publication Status: Available-Online prior to print\n",
      "REFERENCE   2  (bases 1 to 14442)\n",
      "  CONSRTM   NCBI Genome Project\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (04-OCT-2014) National Center for Biotechnology\n",
      "            Information, NIH, Bethesda, MD 20894, USA\n",
      "REFERENCE   3  (bases 1 to 14442)\n",
      "  AUTHORS   Pan,H.C. and Pan,W.J.\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (06-AUG-2014) College of Life Science, Anhui Normal\n",
      "            University, 1 East Beijing road, Wuhu, Anhui 241000, China\n",
      "COMMENT     REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "            reference sequence is identical to KM272950.\n",
      "            COMPLETENESS: full length.\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..14442\n",
      "                     /organism=\"Oxyopes sertatus\"\n",
      "                     /organelle=\"mitochondrion\"\n",
      "                     /mol_type=\"genomic DNA\"\n",
      "                     /specimen_voucher=\"2013-phc-nj-ox-se\"\n",
      "                     /db_xref=\"taxon:93706\"\n",
      "                     /country=\"China: Nanjing\"\n",
      "     tRNA            1..50\n",
      "                     /product=\"tRNA-Met\"\n",
      "     gene            50..1006\n",
      "                     /gene=\"ND2\"\n",
      "                     /db_xref=\"GeneID:20523569\"\n",
      "     CDS             50..1006\n",
      "                     /gene=\"ND2\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 2\"\n",
      "                     /protein_id=\"YP_009072489.1\"\n",
      "                     /db_xref=\"GeneID:20523569\"\n",
      "                     /translation=\"MLYSNFILFVFMYLMSFHLVMSSDDWFLVWLGLEINMMMFIMIVY\n",
      "                     ERYSVLNIESCMKYFFIQSMGSAILLSLFYFGKMYFKEVVCLVMSYKVGAGPFFFWFPS\n",
      "                     ICGGISWLSCFILMCFQKIIPFILLSMFVSGLMFFVIMLSLAFGIMGSFNQNDLKQLIA\n",
      "                     FSSVYHLGWMMLSLMSSDEFWVIYLIMYSLLIFPMILFFKSKMMISLMDVVSMKYKKWF\n",
      "                     VMMMLSMAGMPPLLGFFLKWFAFITIFLKSYFFWLILMMSSVVMFYVYYRVIYDVLMGF\n",
      "                     FMNGVWTNWMEKSNEVWWDLVTIFGMVMGGSVIMLFC\"\n",
      "     tRNA            1008..1070\n",
      "                     /product=\"tRNA-Trp\"\n",
      "     tRNA            complement(1058..1114)\n",
      "                     /product=\"tRNA-Tyr\"\n",
      "     tRNA            complement(1113..1157)\n",
      "                     /product=\"tRNA-Cys\"\n",
      "     gene            1158..2697\n",
      "                     /gene=\"COX1\"\n",
      "                     /db_xref=\"GeneID:20523557\"\n",
      "     CDS             1158..2697\n",
      "                     /gene=\"COX1\"\n",
      "                     /note=\"TAA stop codon is completed by the addition of 3' A\n",
      "                     residues to the mRNA\"\n",
      "                     /codon_start=1\n",
      "                     /transl_except=(pos:2697,aa:TERM)\n",
      "                     /transl_table=5\n",
      "                     /product=\"cytochrome c oxidase subunit I\"\n",
      "                     /protein_id=\"YP_009072490.1\"\n",
      "                     /db_xref=\"GeneID:20523557\"\n",
      "                     /translation=\"MLRWFYSTNHKDIGTLYLMFGVWSAMVGTAMSVLIRMELGHSGSM\n",
      "                     LGDDHLYNVIVTAHAFVMIFFMVMPILIGGFGNWLVPLMLGAPDMSFPRMNNLSFWLLP\n",
      "                     PSLFLLFMSSMVETGVGAGWTVYPPLASTTGHMGSSMDFAIFSLHLAGASSIMGAVNFI\n",
      "                     STIINMRLSSMSMEKVPLFVWSVFITAILLLLSLPVLAGAITMLLTDRNFNTSFFDPAG\n",
      "                     GGDPILFQHLFWFFGHPEVYILILPGFGIVSHVISSSVGKREPFGSLGMIYAMVGIGGM\n",
      "                     GFVVWAHHMFSVGMDVDTRAYFTAATMIIAVPTGIKVFSWMATIHGSYFKMSTSMMWCI\n",
      "                     GFVFLFTLGGITGVVLSNSSLDVVLHDTYYVVAHFHYVLSMGAVFAILAGITYWFPLFF\n",
      "                     GVVLKEKMTKLHFFVMFLGVNLTFFPQHFLGLNGMPRRYSDYPDAFIFWNMVSSMGSIL\n",
      "                     SLFAVMLFMYIVWEAMILKLFGNNEYFVSSSLEWINNIPPLDHTFNQLNLLSEET\"\n",
      "     gene            2698..3363\n",
      "                     /gene=\"COX2\"\n",
      "                     /db_xref=\"GeneID:20523558\"\n",
      "     CDS             2698..3363\n",
      "                     /gene=\"COX2\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"cytochrome c oxidase subunit II\"\n",
      "                     /protein_id=\"YP_009072491.1\"\n",
      "                     /db_xref=\"GeneID:20523558\"\n",
      "                     /translation=\"MPTWSSLLFQDGVSPVMEHLIFFHDHIMMVMIMIMIMVGYVLVNS\n",
      "                     CFNKFYNLGMFQGQGLESIWTAMPAVFLLFIAFPSLHLLYLMEEGASYDITVKVMGHQW\n",
      "                     YWSYEYSDLGFKSFDSYMLSGDDVMFRLLNVDNSLVIPYNSNVRMIISSMDVIHSWTIP\n",
      "                     SLGVKADAVPGRLNQLFYIFNRVGLFSGQCSEICGVNHSFMPIMIMVVPRMEFLTKWF\"\n",
      "     tRNA            3363..3406\n",
      "                     /product=\"tRNA-Lys\"\n",
      "     tRNA            3407..3455\n",
      "                     /product=\"tRNA-Asp\"\n",
      "     gene            3456..3614\n",
      "                     /gene=\"ATP8\"\n",
      "                     /db_xref=\"GeneID:20523559\"\n",
      "     CDS             3456..3614\n",
      "                     /gene=\"ATP8\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"ATP synthase F0 subunit 8\"\n",
      "                     /protein_id=\"YP_009072492.1\"\n",
      "                     /db_xref=\"GeneID:20523559\"\n",
      "                     /translation=\"MPQLMPMFWKSSVFMGFMLLLVMVVFYFFKKMEDFMMKEDMKLKM\n",
      "                     IKFEFLW\"\n",
      "     gene            3614..4276\n",
      "                     /gene=\"ATP6\"\n",
      "                     /db_xref=\"GeneID:20523560\"\n",
      "     CDS             3614..4276\n",
      "                     /gene=\"ATP6\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"ATP synthase F0 subunit 6\"\n",
      "                     /protein_id=\"YP_009072493.1\"\n",
      "                     /db_xref=\"GeneID:20523560\"\n",
      "                     /translation=\"MSLFSVFDPTSYFGLSLNWLIVFMILLYFPMKFYLIESGLMLIFK\n",
      "                     ILFVSVSKVFKEVSIPNHMALNFLCVMTFMYLVFSNMLGLFPFIFTMTAHPLITLGLGV\n",
      "                     VFWLSFFLMGFSFSFKNSCAHLVPEGSPMVLAPLMVLIESISHLIRPFTLSIRLAANMM\n",
      "                     AGHLIMGLLSSISMISVFGFTSSLILQNILLVLEFGVTVIQGFVFSILLLLYALEYY\"\n",
      "     gene            4280..5065\n",
      "                     /gene=\"COX3\"\n",
      "                     /db_xref=\"GeneID:20523561\"\n",
      "     CDS             4280..5065\n",
      "                     /gene=\"COX3\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"cytochrome c oxidase subunit III\"\n",
      "                     /protein_id=\"YP_009072494.1\"\n",
      "                     /db_xref=\"GeneID:20523561\"\n",
      "                     /translation=\"MEKHYHPFHMVNMSPWPLMSGFGALALVVGLIKLFVFFESDLLFS\n",
      "                     SFYLLVLVALLWWRDVIRESTFQGFHSSVVLKGLMLGMILFIVSEIFFFFSFFWAFFHS\n",
      "                     SLSPAVELGVQWPPMGVFPFNPFQVPLLNTVILLSSGVSVTWAHQMILKEDWESAKMSL\n",
      "                     MVTWMLGLYFLMLQGFEYYMSSFGISDSVFGSTFFMATGFHGFHVIVGSVFLFIIWYRF\n",
      "                     LNYHFSKSHHFGFEAAAWYWHFVDVVWLFLFSVVYWWGT\"\n",
      "     tRNA            5064..5119\n",
      "                     /product=\"tRNA-Gly\"\n",
      "     gene            5116..5472\n",
      "                     /gene=\"ND3\"\n",
      "                     /db_xref=\"GeneID:20523562\"\n",
      "     CDS             5116..5472\n",
      "                     /gene=\"ND3\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 3\"\n",
      "                     /protein_id=\"YP_009072495.1\"\n",
      "                     /db_xref=\"GeneID:20523562\"\n",
      "                     /translation=\"MLYLVIFESIFLVFVVYFLFLMVFLKLSSLVEVMSSYECGFDINS\n",
      "                     LARIVFSYRFFLLAILFIIFDVEISLMLPVPYLMVSYMGVIVFCLFMFILLLGLMYEYF\n",
      "                     FGSLDWLHFYMSKE\"\n",
      "     tRNA            complement(5471..5525)\n",
      "                     /product=\"tRNA-Leu\"\n",
      "     tRNA            5522..5566\n",
      "                     /product=\"tRNA-Asn\"\n",
      "     tRNA            5565..5613\n",
      "                     /product=\"tRNA-Ala\"\n",
      "     tRNA            5610..5657\n",
      "                     /product=\"tRNA-Ser\"\n",
      "     tRNA            5658..5697\n",
      "                     /product=\"tRNA-Arg\"\n",
      "     tRNA            5698..5758\n",
      "                     /product=\"tRNA-Glu\"\n",
      "     tRNA            complement(5743..5791)\n",
      "                     /product=\"tRNA-Phe\"\n",
      "     gene            complement(5791..7424)\n",
      "                     /gene=\"ND5\"\n",
      "                     /db_xref=\"GeneID:20523563\"\n",
      "     CDS             complement(5791..7424)\n",
      "                     /gene=\"ND5\"\n",
      "                     /note=\"TAA stop codon is completed by the addition of 3' A\n",
      "                     residues to the mRNA\"\n",
      "                     /codon_start=1\n",
      "                     /transl_except=(pos:complement(5791..5792),aa:TERM)\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 5\"\n",
      "                     /protein_id=\"YP_009072496.1\"\n",
      "                     /db_xref=\"GeneID:20523563\"\n",
      "                     /translation=\"MNFNSMLTMSISLILLSISMFMLFNNTFLYFTIPMENMLSFQVQL\n",
      "                     SSMLDWMSCLFMSTVLMISSMIMVFSMYYIPKNEHKQFHLMLLMFILSMMILILSNNLF\n",
      "                     LMLLGWDMLGVSSYILVIYYQNWSTASCGSITLLTNRIGDIMILLSISMLMLTLSWELN\n",
      "                     MNEMFPLLILIMLMLASCTKSAQFPFSAWLPMAMSAPTPISALVHSSTLVTAGIYLMIR\n",
      "                     ITTNIHPFILISLMFISSITALYAGLSANWEQDLKKIIALSTLSQMAMMMFAISIKSSS\n",
      "                     LSYFHLIIHAYFKSLMFMCAGILIHESSYQDMRLMSNSNYNPSINSTLALTNMALMGLP\n",
      "                     FTSGFFSKDNIIEFMINSKLETTLSLMMILSIGMTSSYSLRMMYLSSKLLIKSKPDSFN\n",
      "                     HTSIYSNSPMMMMSIMPLILGTSMLWMMKPQNMFIFPNHYKFLITITLIMGFMTGMLLS\n",
      "                     FKSFIYKSMGQFSLSLWFTHIISTKLFNRWFPMMIISFNNDKNWQEMYGPNNSFKQTMK\n",
      "                     SIMIPESTKSSTLMMLLMISLIPIMMI\"\n",
      "     tRNA            complement(7424..7473)\n",
      "                     /product=\"tRNA-His\"\n",
      "     gene            complement(7479..8759)\n",
      "                     /gene=\"ND4\"\n",
      "                     /db_xref=\"GeneID:20523564\"\n",
      "     CDS             complement(7479..8759)\n",
      "                     /gene=\"ND4\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 4\"\n",
      "                     /protein_id=\"YP_009072497.1\"\n",
      "                     /db_xref=\"GeneID:20523564\"\n",
      "                     /translation=\"MLKFIIPLSILFMLQTWLPFLIMSISILSLLIILMNFNSAITIFN\n",
      "                     KMMFQDKSSYILIMLSLMSVMLIILSSLNLKDIFTTIFPILLILILTFSTSNLISFYIL\n",
      "                     FELVLIPTMMLILKMSKQQERLQASIYLMMYTITASLPLLINMILLKNNYSFNIMNIFL\n",
      "                     FKLNMIIMFMMAFLVKTPMFFTHLWLPKAHVEAPLEGSMILAAMLLKLGGYGLIRFIPL\n",
      "                     CIKKMNSINNWIISISLIGASLTSLNCIRQKDLKSLIAYSSVAHMGFVLLGLFSMTKLG\n",
      "                     MFGAILMMIAHGLSSSALFLLTNDLYLKYHSRSIILFKSMITITPNITFWWFMFTMVNI\n",
      "                     SAPPTINTISEIFIMMSTLMWNKMTMIMIFISSIAVTAFSMMMFLNISHNKMHSLPLIH\n",
      "                     TPYKIYLSLFIHLIPLIMLLTKMEILI\"\n",
      "     gene            complement(8760..9027)\n",
      "                     /gene=\"ND4L\"\n",
      "                     /db_xref=\"GeneID:20523565\"\n",
      "     CDS             complement(8760..9027)\n",
      "                     /gene=\"ND4L\"\n",
      "                     /note=\"TAA stop codon is completed by the addition of 3' A\n",
      "                     residues to the mRNA\"\n",
      "                     /codon_start=1\n",
      "                     /transl_except=(pos:complement(8760),aa:TERM)\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 4L\"\n",
      "                     /protein_id=\"YP_009072498.1\"\n",
      "                     /db_xref=\"GeneID:20523565\"\n",
      "                     /translation=\"MSMIPTMVIISILSMCWWRKNILIMLLSLELLLISLFSFISTNFS\n",
      "                     SISFPSMMIMLTLMVSGSSMSLSMLVSLSNLHNSSNSVFISSLT\"\n",
      "     tRNA            complement(9028..9076)\n",
      "                     /product=\"tRNA-Pro\"\n",
      "     gene            9084..9521\n",
      "                     /gene=\"ND6\"\n",
      "                     /db_xref=\"GeneID:20523566\"\n",
      "     CDS             9084..9521\n",
      "                     /gene=\"ND6\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 6\"\n",
      "                     /protein_id=\"YP_009072499.1\"\n",
      "                     /db_xref=\"GeneID:20523566\"\n",
      "                     /translation=\"MNLIFLLGVFFVVSLQPVFMVSILIFITLFYSYLIYKVIGTFWFS\n",
      "                     YMLLMVMLSGVLVIFTYMVSLVPNESFEVVNLVILLSFMISMFVKWGLDFGGDMGIMSM\n",
      "                     DLWISYLGGFSLFLVVFLLVIMLLVVVVSCMSGMGSFRVSW\"\n",
      "     tRNA            9515..9568\n",
      "                     /product=\"tRNA-Ile\"\n",
      "     gene            9566..10674\n",
      "                     /gene=\"CYTB\"\n",
      "                     /db_xref=\"GeneID:20523567\"\n",
      "     CDS             9566..10674\n",
      "                     /gene=\"CYTB\"\n",
      "                     /note=\"TAA stop codon is completed by the addition of 3' A\n",
      "                     residues to the mRNA\"\n",
      "                     /codon_start=1\n",
      "                     /transl_except=(pos:10673..10674,aa:TERM)\n",
      "                     /transl_table=5\n",
      "                     /product=\"cytochrome b\"\n",
      "                     /protein_id=\"YP_009072500.1\"\n",
      "                     /db_xref=\"GeneID:20523567\"\n",
      "                     /translation=\"MFMSLRKENKLLMIMNNSLVDLPSPSSLSYFWNFGSLLGVFLMVQ\n",
      "                     ILSGLFLSFHFSGDVNLSFFSIIHMMRDVNYGWLLRAIHANGASMFFLLMYIHIGRGLY\n",
      "                     YGSYRFEKVWLSGVSILLLSMATAFLGYVLPWGQMSFWAATVITNLLSAIPYVGVLLVE\n",
      "                     WIWGGFSVGNPTLTRFFAFHFILPFVILFFVILHLFFLHEKGSSNPLGLMGSSDKIMFH\n",
      "                     PYYVFKDIFGFSFFFMFFMVICLLYPYIFMDVENFISSNPLVTPVHIQPEWYFLFAYTI\n",
      "                     LRSVPSKIGGVVALVMSVLVLYFLPFFLKHRFRSTFFYFLMKFLFWIFFVNWVLLTWIG\n",
      "                     ACVVEIPYMTLGMIFSVIYYMMYIMFWMV\"\n",
      "     tRNA            10688..10755\n",
      "                     /product=\"tRNA-Ser\"\n",
      "     tRNA            10756..10807\n",
      "                     /product=\"tRNA-Thr\"\n",
      "     gene            complement(10805..11723)\n",
      "                     /gene=\"ND1\"\n",
      "                     /db_xref=\"GeneID:20523568\"\n",
      "     CDS             complement(10805..11723)\n",
      "                     /gene=\"ND1\"\n",
      "                     /note=\"TAA stop codon is completed by the addition of 3' A\n",
      "                     residues to the mRNA\"\n",
      "                     /codon_start=1\n",
      "                     /transl_except=(pos:complement(10805),aa:TERM)\n",
      "                     /transl_table=5\n",
      "                     /product=\"NADH dehydrogenase subunit 1\"\n",
      "                     /protein_id=\"YP_009072501.1\"\n",
      "                     /db_xref=\"GeneID:20523568\"\n",
      "                     /translation=\"MLFSMNFILISISILVAVAFYTILERKILSYTQIRKGPNKVGLIG\n",
      "                     LLQPFSDAIKLFNKNLISLESMNFMLSYLTPSLSLLIMMMMVPIILYNSNSLMDNKHNM\n",
      "                     LYFFILSSLGVYFILLVGWSSNSKYSYLGAIRSIAQMISYEVSFFMIILFIVIMSQSYS\n",
      "                     FTQMEESQKLLHFFWGNMFLMFMWLTSCLAETNRSPFDFAEGESELVSGFNVEYMGGWF\n",
      "                     ALIFLAEYMSMLLLSMISTIMFFKNIKSMFSMFILLVISVILIWVRSTFPRFRYDMLMK\n",
      "                     LSWKIFLPMSMLTIIIPIFTFYMFI\"\n",
      "     tRNA            complement(11724..11773)\n",
      "                     /product=\"tRNA-Leu\"\n",
      "     rRNA            complement(11774..12794)\n",
      "                     /product=\"16S ribosomal RNA\"\n",
      "     tRNA            complement(12795..12844)\n",
      "                     /product=\"tRNA-Val\"\n",
      "     rRNA            complement(12845..13541)\n",
      "                     /product=\"12S ribosomal RNA\"\n",
      "     tRNA            complement(13542..13603)\n",
      "                     /product=\"tRNA-Gln\"\n",
      "ORIGIN\n",
      "        1 attaatataa gctaattaag ctaatgagtt catatctcat aaatgattaa ttttgtattc\n",
      "       61 taatttcatt ctttttgtat ttatatattt aataagattt catttagtaa taagatctga\n",
      "      121 tgattgattt ttagtttgat taggattaga aattaatata atgatattta ttataattgt\n",
      "      181 ttatgaacgg tatagagtat taaatattga gtcttgtata aaatattttt ttatccaaag\n",
      "      241 aataggctct gctattttat tgagtttatt ttattttgga aaaatatatt ttaaggaagt\n",
      "      301 agtttgttta gtaataagat ataaagttgg agcaggaccc tttttttttt ggtttccttc\n",
      "      361 tatttgtggt gggatttctt gattatcttg ttttatttta atatgttttc aaaagattat\n",
      "      421 tccttttatt ttattaagaa tgtttgtatc tggtttgata ttttttgtaa ttatacttag\n",
      "      481 tttggctttt ggaattatag gaagatttaa tcaaaatgat ttaaagcaat taattgcttt\n",
      "      541 ttcttcagtg tatcatttag gatgaataat attaagatta atatcttctg atgaattttg\n",
      "      601 ggtaatttat ttaattatat atagattatt aatttttcct ataattcttt tttttaaaag\n",
      "      661 aaagataata attagattaa tagatgtagt taggataaaa tataaaaagt gatttgtaat\n",
      "      721 aatgatatta agaatagcag gcatacctcc cttattagga ttttttttaa agtgatttgc\n",
      "      781 ttttattacg atttttttaa aaagatattt tttttgatta attttaatga tgagttctgt\n",
      "      841 ggtaatgttt tatgtatatt atcgtgttat ttatgatgtt ttgatgggat tttttataaa\n",
      "      901 tggggtttgg acaaattgaa tggaaaaaag aaatgaggtt tggtgagatt tagtaactat\n",
      "      961 ttttggtata gttataggtg gttcggtaat tatattattt tgttagtata atatatatat\n",
      "     1021 atatatagga aagtaagata tggtctatta attttcaagg ttaaaagtgt atactaatat\n",
      "     1081 gaatttacag ttcatcatca atgatttctt aatttcttta aatttgcaat ttaatgtttt\n",
      "     1141 ataaactata tagagatata ctgcgatgat tttattcaac taatcataag gatattggaa\n",
      "     1201 ctttatattt aatatttggg gtttgatcgg ctatagttgg aacggcaata agagtattga\n",
      "     1261 ttcggatgga attagggcat tctgggagaa tattagggga tgatcattta tataatgtga\n",
      "     1321 ttgttactgc tcatgctttt gtaataattt tttttatagt tataccaatt ttaattggtg\n",
      "     1381 ggtttggaaa ttgattggtt cctttgatat taggggctcc tgatatgtct ttccctcgaa\n",
      "     1441 taaataattt atcgttttga ttacttcctc cttctttgtt tttattattt atatcttcta\n",
      "     1501 tggttgaaac tggggttgga gcaggatgaa cggtttatcc tcctttagct tcgactactg\n",
      "     1561 gtcatatagg aagatcaata gattttgcta ttttttcttt acatttagct ggtgcttctt\n",
      "     1621 ctattatagg ggctgtaaat tttatttcta ctattattaa tatacgatta agtagaataa\n",
      "     1681 gaatggaaaa ggttcctttg tttgtctgat ctgtttttat tactgctatt ttattgttgt\n",
      "     1741 tatctttacc agtattagct ggagctatta ctatgttatt aactgatcga aattttaata\n",
      "     1801 cttcattttt tgatcctgca ggggggggag atcctatttt atttcaacat ttattttgat\n",
      "     1861 tttttggaca tccagaagta tatattttaa ttttacctgg atttggaatt gtttctcatg\n",
      "     1921 taattagatc ttctgtggga aagcgagaac cgtttggaag attaggaata atttatgcta\n",
      "     1981 tagtaggaat tgggggtata ggatttgttg tttgagcaca tcatatattt tcggtaggaa\n",
      "     2041 tagatgtaga tactcgagct tattttactg cggctactat aattattgct gtcccaactg\n",
      "     2101 ggattaaggt ttttagttga atagcaacaa ttcatggttc ttattttaaa ataagaactt\n",
      "     2161 ctataatatg atgtattggg tttgtatttt tatttacttt aggtggtatt actggggttg\n",
      "     2221 ttttatctaa ttcatcgtta gatgtagtat tacatgatac atattatgta gtagctcatt\n",
      "     2281 ttcattatgt attaagaata ggagctgtat ttgctatttt agctggaatt acttattgat\n",
      "     2341 ttcctttatt ttttggagta gttttgaagg agaaaatgac taagcttcat ttctttgtta\n",
      "     2401 tgtttttagg agtaaatttg acattctttc ctcaacattt tttaggatta aatgggatac\n",
      "     2461 cacgtcgata ttcagattat cctgatgctt ttattttttg aaatatagtt tcttctatag\n",
      "     2521 gttctatttt atctttattt gctgttatac tatttatata tattgtttga gaagctataa\n",
      "     2581 ttttaaaatt atttgggaat aatgaatatt ttgtgagttc ttcgttagaa tgaattaata\n",
      "     2641 atattcctcc tttagatcat acttttaatc aattgaattt attaagagaa gaaacttttg\n",
      "     2701 ccaacttgaa gatctttgtt gtttcaagat ggagtttctc cagtaataga acatttaatt\n",
      "     2761 ttttttcatg atcatattat aatagtaata attataatta taattatggt aggatatgta\n",
      "     2821 ttagtgaatt cttgttttaa taaattttat aatttaggaa tgtttcaagg tcaggggctt\n",
      "     2881 gaaagaattt gaacagcaat acctgctgta tttttgttat ttattgcttt tccttcttta\n",
      "     2941 catttgttat atttgatgga agaaggagct tcttatgata ttacggttaa ggttatagga\n",
      "     3001 caccaatggt attggtctta tgaatataga gatttaggat ttaagtcttt tgattcatat\n",
      "     3061 atattatcag gtgatgatgt aatatttcga ttgttaaatg ttgataattc attagttatt\n",
      "     3121 ccttataata gaaatgttcg aataattatt tcaagaatag atgtaattca ttcttgaact\n",
      "     3181 attccttctt taggtgtaaa ggctgatgcc gtacctggtc gattaaatca attgttttat\n",
      "     3241 atttttaatc gtgtaggatt atttagtgga caatgttcag aaatttgtgg tgtaaatcat\n",
      "     3301 agatttatac ctattataat tatagttgtt cctcgaatag agtttttgac aaagtgattt\n",
      "     3361 taagatagtg gccgaattta ggtgttagtc tcttaaatta attatgaata gtttagttaa\n",
      "     3421 taataatatt aatttgtcaa gttaaagtaa ggaaaattcc tcagttaata ccaatatttt\n",
      "     3481 gaaaaagatc tgtttttata ggttttatat tattattagt aatagtagtc ttctattttt\n",
      "     3541 ttaaaaaaat agaagacttt ataataaagg aagatataaa gttaaaaatg attaagtttg\n",
      "     3601 aatttttatg ataataagtt tattttcggt atttgatcct acttcttatt ttggcttaag\n",
      "     3661 attaaattga ttaattgttt ttataatttt attgtatttt cctataaagt tttatttaat\n",
      "     3721 tgaaagtgga ttaatattaa tttttaaaat tttatttgtt agagtaagta aggtttttaa\n",
      "     3781 agaagtttct attccaaatc atatagcatt aaatttttta tgtgtaataa cgtttatata\n",
      "     3841 tttagtgttt agtaatatac taggtttatt tccatttatt tttactataa cggctcatcc\n",
      "     3901 attaattaca ttaggattag gagttgtttt ttggttgtca ttttttttaa taggattttc\n",
      "     3961 atttagattt aaaaatagat gtgctcatct agtaccagaa ggtagtccta tagttttagc\n",
      "     4021 tcctttaata gtattaattg agagaatcag tcatttaatt cgacctttta ctctttctat\n",
      "     4081 tcgattagcg gctaatataa tggcgggcca tttaattata ggattattat ctagaattag\n",
      "     4141 aatgattagt gtatttggtt ttactagatc tttaatttta caaaatattc ttttggtgtt\n",
      "     4201 agaatttgga gtaacagtta ttcaaggatt tgtttttaga attttattat tattatatgc\n",
      "     4261 attggaatat tattaacttt tggaaaagca ttatcatcct tttcatatag taaatatatc\n",
      "     4321 tccttgacct ttgataagag gatttggagc gttggcttta gtagttggat taattaaact\n",
      "     4381 ttttgttttt tttgaaagag atttattatt tagatcattt tatttattag tattagtagc\n",
      "     4441 gttattatga tgacgtgatg ttattcggga aagtaccttt caaggttttc attctagtgt\n",
      "     4501 ggtgttgaaa ggacttatat tagggataat tctatttatt gtaagagaaa tttttttttt\n",
      "     4561 tttttctttt ttttgagctt tttttcattc tagattgaga ccggcagtag aattaggagt\n",
      "     4621 tcagtgacca cctatgggtg tgtttccttt taatcctttt caagttcctt tattaaatac\n",
      "     4681 ggttatttta ttaagatctg gtgtgagagt tacttgagct catcaaatga ttttaaagga\n",
      "     4741 ggattgagaa agagctaaga tgtctttaat agttacgtga atattgggtt tatattttct\n",
      "     4801 tatgttgcaa ggttttgagt attatatatc ttcttttgga attagagatt ctgtatttgg\n",
      "     4861 ttctactttt tttatagcga ctggatttca tggatttcat gtaattgtgg gaagagtgtt\n",
      "     4921 tttatttatt atttgatatc ggtttttaaa ttatcatttt tctaaaagtc atcattttgg\n",
      "     4981 gtttgaagct gctgcttgat attgacattt tgtggatgtg gtttggttgt ttttgttttc\n",
      "     5041 tgtagtatat tgatgaggaa cttaaagagt aagaagtatt tttgtatgtt taatttccaa\n",
      "     5101 ttaaaaggtg aatttatttt atatttggtt atttttgaat ctatcttttt agtgtttgtg\n",
      "     5161 gtttattttt tgtttttaat agttttttta aaattaagaa gattagtgga agtaataaga\n",
      "     5221 tcatatgagt gtggatttga tattaattct ttagcacgta ttgtgttttc ttatcgattt\n",
      "     5281 tttttattag ctattttgtt tattattttt gatgtagaaa tttctttgat attgccggtt\n",
      "     5341 ccctatttga tagttagata cataggggta attgtttttt gtttgtttat atttattctg\n",
      "     5401 ttattaggtt taatatatga atattttttt ggatctttag attgattaca tttttatata\n",
      "     5461 tctaaggagt aaaatttatt tttaagactt aaacttaatg catttgttct gccaatagaa\n",
      "     5521 ttgttaaata gagctgtaat agcgttttca ttgttaatga aatgaaggaa gttttgaatt\n",
      "     5581 gatttgcaat caattaatat ggtaacatac ggaaggttta attaaaagct gctaactttt\n",
      "     5641 atatagctaa atgctacaat taggagcgta agcgacttaa tttcgactta agttaaggat\n",
      "     5701 taagaagttt tataacgtca ttgtttcatg atggagatgt gttagtcctc tttatcttca\n",
      "     5761 gtaaagtgct ttgagaatca gctaattaaa taaattatta taattggaat taatgaaatt\n",
      "     5821 attaataata ttattaaagt agatgatttt gtagattctg gaattataat tctttttatt\n",
      "     5881 gtttgtttga atgagttatt tggaccatat atttcttgtc agtttttgtc gttattaaat\n",
      "     5941 gagatgatta ttattggaaa tcaacggtta aataactttg ttgaaataat atgggtgaat\n",
      "     6001 cataatgata aagaaaattg acctattgat ttgtagataa atcttttaaa agatagtaat\n",
      "     6061 attcctgtta taaaacctat aattaatgta attgtaatta aaaatttata atggtttggg\n",
      "     6121 aaaatgaata tattttgggg ttttattatt cataatattg aggttcctaa aattaaaggt\n",
      "     6181 ataattctta ttattattat tggggaattt gaataaattg aagtgtgatt aaaagaatct\n",
      "     6241 ggtttggatt taattaaaag tttacttgaa agatatatta ttcgtaagga gtacgatgaa\n",
      "     6301 gttattccga tagataaaat tattattaat gataatgttg tttctaattt tgaattaatt\n",
      "     6361 ataaattcaa taatattatc tttagaaaag aatcctgagg taaatggaag tcctattaga\n",
      "     6421 gctatattag ttaaagctaa tgtggaatta atagatggat tataatttga attagatatt\n",
      "     6481 aggcgtatat cttgatagga agattcatga attaaaattc cagcacatat aaatattagt\n",
      "     6541 gatttgaaat aagcgtgaat aataaggtgg aaatatctta gagaacttga tttaatggaa\n",
      "     6601 attgcaaata ttattattgc tatttgtctt agggtggata aagcaataat tttttttaaa\n",
      "     6661 tcttgttccc agttagctga aagtccagca tataatgctg taattgaaga aatgaatatt\n",
      "     6721 aaggaaatta aaataaaagg gtgaatatta gttgtaattc ggattatcaa ataaattcct\n",
      "     6781 gcagttacaa gagtggagga atgaactagt gctgaaatag gagttggagc tgatattgct\n",
      "     6841 atgggaagtc aagctgagaa aggaaattga gcacttttag tgcatctagc tagtattaat\n",
      "     6901 atgattaaaa taagtaatgg aaatatttca tttatgttta gttcccatct taatgttaat\n",
      "     6961 attaatattc taattgataa aagaattata atgtctccga ttcgatttgt gaggagtgta\n",
      "     7021 attgaaccac atctagctgt ggatcagttt tgataataga ttactaaaat ataagaagat\n",
      "     7081 actcctagta tatctcatcc taataatatt aaaaatagat tgtttcttaa aattaagatt\n",
      "     7141 attatggata agatgaatat taatagtatt aaatgaaatt gtttgtgttc atttttaggg\n",
      "     7201 atataatata ttctaaatac tataattata cttgaaatta taagtactgt tcttatgaat\n",
      "     7261 agacaggata ttcaatctag tattgaagat aattgtactt gaaatgataa tatgttttct\n",
      "     7321 atagggattg taaaatacaa aaatgtatta ttaaataata taaatattct gattgaaagt\n",
      "     7381 aaaattaaag aaattgatat tgttagtata gaattaaaat ttataaaaga atttataact\n",
      "     7441 ccacaagtta ttgttttatt taaactagaa tattatattt aaataaggat ttctattttt\n",
      "     7501 gttaataata taataagtgg aattaaatgg ataaataatg ataagtagat tttgtaaggt\n",
      "     7561 gtgtgaatta aagggagaga atgtatttta ttatgagaga tattgagaaa tattattatt\n",
      "     7621 gaaaatgcgg ttactgcgat tgaagaaata aaaattataa ttattgttat tttgtttcat\n",
      "     7681 attaatgttc ttattataat aaaaatttct ctaattgtat tgatagtagg gggagctgaa\n",
      "     7741 atatttacta tagtaaatat aaatcatcaa aaagtaatgt taggggtaat tgtaattatt\n",
      "     7801 cttttaaata aaatgattct tcgggaatgg tattttaaat ataagtcatt tgttaataaa\n",
      "     7861 aataatgctg atgatgataa accatgagcg attattatca aaattgctcc aaatattcct\n",
      "     7921 agttttgtta tcgaaaataa tcctagtaga acgaatccta tatgggccac tgaggagtaa\n",
      "     7981 gcaattaatg atttaagatc tttttggcga atacagttta atcttgttaa ggaagcaccg\n",
      "     8041 attaatctga ttctaataat tcaattgtta attgaattta tttttttaat acataatgga\n",
      "     8101 ataaatcgaa ttaatccata tcctcctaat tttaataata tagctgctaa aattattgat\n",
      "     8161 ccttctaggg gtgcttctac atgggctttt ggtaatcata aatgggtaaa aaatattggg\n",
      "     8221 gtttttacga gaaaagctat tataaatata ataattatat ttaatttaaa taaaaaaatg\n",
      "     8281 tttataatat taaatgagta attgtttttt aatagaatta tattgataag gagaggtaag\n",
      "     8341 gatgctgtga ttgtatatat tattagatag attctggctt ggagtcgttc ttgttgtttt\n",
      "     8401 cttattttta agattaatat tatagtagga attaatacta attcaaatag aatataaaat\n",
      "     8461 gaaattaaat tagaggtgga gaaagttaaa attaagatta aaagaattgg aaagattgtt\n",
      "     8521 gtaaaaatgt cttttagatt taatgaggat agaataatta atatcactct tattaatgat\n",
      "     8581 agtataatta aaatatagga tgatttgtcc tgaaatatta ttttgttaaa aatagtaatt\n",
      "     8641 gctgaattga agtttattaa aataattagt aaagataaaa tggaaataga tataattaag\n",
      "     8701 aatggaagtc atgtttgtaa tataaataag atagaaaggg gaataataaa ttttagcaaa\n",
      "     8761 agttaaagat ctaataaaaa ctgagttgga agaattgtgt aaatttgata gtgatactaa\n",
      "     8821 tattcttagt cttattcttg aacctcttac tattaatgta agtataatta ttattgatgg\n",
      "     8881 aaatgaaatt gatgaaaagt tagttgaaat gaatgaaaat aaagaaatta ataataattc\n",
      "     8941 taatcttaga agtataatta aaatattttt tcgtcatcaa catattctta agattctgat\n",
      "     9001 aataactata gttggaatta ttgacattta agaattttct acatccaaag tagatatttt\n",
      "     9061 agttaaatta tttagaactt ttggtgaatt tgattttttt attgggtgta ttctttgtag\n",
      "     9121 ttagattaca accggtattt atggtaagaa ttttaatttt tattacttta ttttattctt\n",
      "     9181 atttgattta taaggtaatt ggaacatttt ggtttagata tatgttattg atagtaatat\n",
      "     9241 taagaggggt tttagtaatt tttacttata tagtaagatt agttcctaat gagagttttg\n",
      "     9301 aagttgttaa tttagtgatt ttattaagat ttataattag tatatttgtt aaatgggggt\n",
      "     9361 tagattttgg cggagatata ggaattataa gaatggattt gtgaatttcg tatttaggag\n",
      "     9421 gttttagatt gtttttggtt gtttttcttt tggtaattat attgttagtg gtggtagtga\n",
      "     9481 gatgtatgag tggaatggga agttttcgag taagatgata aagtgcctga ttaaaggact\n",
      "     9541 aatttgatag attagaacat gattgatttt tatatcttta cgaaaggaaa ataagttatt\n",
      "     9601 gataattata aataattctt tagttgattt accttctcca tctagtttaa gatatttttg\n",
      "     9661 aaattttggt tctttgttag gtgtattttt aatggtgcag attttatctg gtttattttt\n",
      "     9721 atcttttcat tttagaggag atgtaaattt gtcttttttt agtattattc atataatacg\n",
      "     9781 agatgtaaat tatggatgat tgttacgagc tattcatgct aatggtgcta gaatgttttt\n",
      "     9841 tttattgata tatattcata ttggtcgggg attgtattat ggttcatatc gatttgagaa\n",
      "     9901 ggtttgattg agaggggtta gaattttatt attatctatg gctacggctt ttttagggta\n",
      "     9961 tgttttgcct tgggggcaaa tgtctttttg agctgctaca gtgattacta atttattatc\n",
      "    10021 ggctattcct tatgtaggag tattattggt agagtgaatt tgagggggat tttcggtagg\n",
      "    10081 aaatcctact ctgactcgat tttttgcttt tcattttatt ttaccttttg taattttgtt\n",
      "    10141 ttttgttatt ctgcatttgt tttttcttca tgagaaagga tctagaaatc ctttaggatt\n",
      "    10201 aatgggtaga agagataaaa ttatatttca tccttattat gtttttaagg atatttttgg\n",
      "    10261 atttagtttt tttttcatgt tttttatagt gatttgtttg ttgtatcctt atatttttat\n",
      "    10321 ggatgtggaa aattttattt cttctaatcc tttagttacg cctgttcata ttcaacctga\n",
      "    10381 atgatatttt ttgtttgctt atactattct tcgttctgta ccaagaaaga ttgggggagt\n",
      "    10441 tgttgcgttg gtgatatcag tattagtttt gtattttttg cctttttttt taaaacatcg\n",
      "    10501 ttttcgaaga actttttttt attttttaat aaaattttta ttttgaattt tttttgtaaa\n",
      "    10561 ttgagtgctt ttaacttgaa ttggggcttg tgtagttgaa attccttata tgacattagg\n",
      "    10621 aatgattttt agagtaattt attatatgat atatattatg ttttgaatag tatatgaaat\n",
      "    10681 gcaagatctt gtaatgttat agactttata taaagttgtt ttgaaaacaa tttataagat\n",
      "    10741 tattctttaa agtcaggttg attagtttaa gtaaaacgta aattttgtaa gtttaatttc\n",
      "    10801 tataaaataa atatatagaa agtgaaaata gggataataa ttgttagtat agatataggt\n",
      "    10861 aaaaagattt ttcatcttag ttttattaat atatcatatc ggaatcgggg aaatgttctt\n",
      "    10921 cgtactcaaa ttaaaataac tgaaataact aataagataa atattgaaaa tattgattta\n",
      "    10981 atatttttaa aaaatataat agtagaaatt attcttaaaa gaagtatact tatatattca\n",
      "    11041 gctaagaaaa ttagtgcgaa tcaacctcct atatattcta cattaaatcc tgatactaat\n",
      "    11101 tctgattctc cttctgcgaa atcgaaagga cttcggttgg tttctgctaa acaagaagtt\n",
      "    11161 aatcatataa atataagaaa tatgtttcct cagaaaaaat gtaagagttt ttgtgattct\n",
      "    11221 tctatttgag taaatgaata agactgtgat ataataacaa taaataagat aattataaaa\n",
      "    11281 aaagatactt cgtaagaaat tatttgagca attcttcgaa ttgccccaag atatgaatat\n",
      "    11341 tttgaattag atgatcatcc tactaataaa atgaaatata ctcctagtct tgataaaata\n",
      "    11401 aagaagtata atatgttgtg tttattgtct attaatgaat ttgaattgta taaaataata\n",
      "    11461 gggactatta ttattataat gagtaaggat aatgatgggg ttaaatatga tagtataaaa\n",
      "    11521 tttattgatt ctaaggagat taaattttta ttaaaaagtt taattgcgtc tctgaatggt\n",
      "    11581 tgaagtaatc caattagtcc tactttattt ggaccttttc ggatttgagt atatctaagg\n",
      "    11641 atttttcgtt ctaaaatagt gtaaaaagct actgctacta aaattctaat tgaaattaga\n",
      "    11701 ataaagttta tggaaaataa tattttagtt taattagatt ctaattctaa tgcattgttt\n",
      "    11761 ctgccaagat aatgattaaa tcattttctt aggtcctttc gtactagaag aaataataag\n",
      "    11821 aagatagaaa ccgacctggt ttacaccggt ctgaactcaa atcatgtaat tttttaagag\n",
      "    11881 tcgaacagac tttctttatc cactattgcg cagataagaa tattaaatca acatcgaggt\n",
      "    11941 cgcaatcatt tttttaaata agatctttta aattttatta cgctgttatc cctacggtaa\n",
      "    12001 tttggtcatt taattaattt tattagatca gttttataag tatttttatt aaaatattaa\n",
      "    12061 ttaattaact gccccagtta aactaaagta aaattcaata gggtcttatc gtctttctgt\n",
      "    12121 tatataaaag gctttttact tttaagataa ttttaatatt taaattaaag aaataatttt\n",
      "    12181 aagtgttaaa tcttttgttc tagtctttaa ttaaaagaca aatgattatg ctaccttagc\n",
      "    12241 acaattgcgg ctattcaaat atttcattga gcagatttta ctattaatta tgattaatag\n",
      "    12301 aaatgttttt gttaaacagt cattaaatat atgagttatt gtaatttatt taaataattt\n",
      "    12361 acttatttta ttatttttat attattaaaa tgatatttta atagtatagt ttaatttaat\n",
      "    12421 ttgtttagtt aagttattaa actatttgaa tattataatt tctatatttg taagtatgtt\n",
      "    12481 ttttaaattt tatttgtaat ttgttttaca aaaattattt ttaatatatt tttttataaa\n",
      "    12541 ataccagata ttattttaat cgaataacgt ctaatttcga ttgaatatta taattattca\n",
      "    12601 tataatgaat ttataatatt aaatagatct taaaatttcg ggaattaaaa aatttttatt\n",
      "    12661 tgaattaata gatcctgata ctaaaggaat attataattc tttgaaaaag tttatttttc\n",
      "    12721 cttttcagtt taaattattt atgtgtaatt aaagaataaa agtaattttt taaaaattag\n",
      "    12781 tttttggaaa tatctttaaa gattcatttc agtgtaagtg aagtgctata aaagcttatt\n",
      "    12841 agttcttgat actttttcca aaataccaac tgtgttacga cttaccttat cttatgataa\n",
      "    12901 gggtgacggg cgatatgtgc acatttatta ttttattcaa atttaaatat tattaaaaat\n",
      "    12961 ttacttctaa atcctatttc ttaaaaatat taaagaataa tgttctttaa ttaaatttta\n",
      "    13021 atgtaactca tttattcttt aacattaatt acaccttgac ctaacatttt ataatttatt\n",
      "    13081 tagaataata ttattaatta gttcttttga tggcggtata taaactttgt tttaaagtaa\n",
      "    13141 aatttatcgg gggttgtcga ttaataaaca agttcctcta gaaaaataaa atgccgccgt\n",
      "    13201 tgtaaataag ttttattata aattattatt actatattta ttttaatata ctagggtttc\n",
      "    13261 taatcctatt ttaagtttaa aaattataac ttttatttaa ataaatttta tataattaga\n",
      "    13321 attttacttt ttagtaattg attttttatt aattttaatt atgttaatat aaatataatt\n",
      "    13381 taactaaaag aataaccgca actgctggca cttgaattat tagttagtta tttagttaat\n",
      "    13441 tattctttat attttaaata atgttagatg ttacaaaatt aaattgattt ttataatttt\n",
      "    13501 gtattaatat aatatattat aaatttttaa taatgtaatt caattagata tgatatatct\n",
      "    13561 ctataaaaat caaattttta tgtgcgaaaa gcactttaat agataatttt aattttaaaa\n",
      "    13621 aaattacttt ttaaaaagag tagtaattaa atttaggtta ttattttgag tggtttggga\n",
      "    13681 aaaatctatt cttttttttt ttttttaaaa aaaaagggta tggtttttga gtcgggtgaa\n",
      "    13741 ttattttgga atgattttta aaatttttgt ggaaaatttt tttaaaggta taaattaatt\n",
      "    13801 taaataccca gattttctat tattttttgt aacttttcaa aatttaaaat ctttactttt\n",
      "    13861 tttttgtaga gatatcggaa aaggctattg agatactaac gtatatatat atatatatat\n",
      "    13921 attcattatt aaagttcttt tggagattga attaaatcaa tagtaatttc tttaaaagct\n",
      "    13981 tatcttttaa agaggataga tttttcaggc ccgataccgg gaaaatcggg cctgggaaaa\n",
      "    14041 taatatctat actcttaata gtagtattta taatgtttct tgatattata atagtgtatt\n",
      "    14101 tgtgtaatta catattgatt tttattgttt ttgttactca ttctaattta tattgttact\n",
      "    14161 atttacttta gaatgagtaa gtgttgttgt tattattatt gtattgttgt tatgattttt\n",
      "    14221 ttttaatagt ttctttgaga ttattttctg tagaggatag atttttcagg cccgataccg\n",
      "    14281 ggaaaatcgg gcctgggaaa ataatatcta tactcttaat agtagtattt ataatgtttc\n",
      "    14341 ttgatattat aatagtgtat ttgtgtaatt acatattgat ttttattgtt tttgttactc\n",
      "    14401 attctaattt atattgttac tatttacttt agaatgagta ag\n",
      "//\n",
      "\n",
      "14442\n"
     ]
    }
   ],
   "source": [
    "print(gb.format(\"gb\"))\n",
    "print(len(gb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14442\n"
     ]
    }
   ],
   "source": [
    "print(len(gb.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Just me messing around with codon tables and such"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Phe': ['TTT', 'TTC'], 'Leu': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'Ser': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'], 'Tyr': ['TAT', 'TAC'], 'Cys': ['TGT', 'TGC'], 'Trp': ['TGA', 'TGG'], 'Pro': ['CCT', 'CCC', 'CCA', 'CCG'], 'His': ['CAT', 'CAC'], 'Gln': ['CAA', 'CAG'], 'Arg': ['CGT', 'CGC', 'CGA', 'CGG'], 'Ile': ['ATT', 'ATC'], 'Met': ['ATA', 'ATG'], 'Thr': ['ACT', 'ACC', 'ACA', 'ACG'], 'Asn': ['AAT', 'AAC'], 'Lys': ['AAA', 'AAG'], 'Val': ['GTT', 'GTC', 'GTA', 'GTG'], 'Ala': ['GCT', 'GCC', 'GCA', 'GCG'], 'Asp': ['GAT', 'GAC'], 'Glu': ['GAA', 'GAG'], 'Gly': ['GGT', 'GGC', 'GGA', 'GGG'], 'End': ['TAA', 'TAG', 'AGA', 'AGG']}\n"
     ]
    }
   ],
   "source": [
    "codon_tables = set_ncbi_genetic_codes()\n",
    "\n",
    "print(codon_tables[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Allochrocebus_lhoesti_NC_023962.1.gb\n"
     ]
    }
   ],
   "source": [
    "print(genbank[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "fields = ['species', 'transl_table', 'aminoacid', \"Anticodon\", \n",
    "                  'Number of codon occurences', '/1000', 'Fraction']\n",
    "excel_output_dict = {k : list() for k in fields}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTT 0.45\n",
      "TTC 0.55\n",
      "TTA 0.16\n",
      "TTG 0.03\n",
      "CTT 0.15\n",
      "CTC 0.16\n",
      "CTA 0.45\n",
      "CTG 0.04\n",
      "TCT 0.19\n",
      "TCC 0.29\n",
      "TCA 0.28\n",
      "TCG 0.03\n",
      "AGT 0.06\n",
      "AGC 0.14\n",
      "TAT 0.4\n",
      "TAC 0.6\n",
      "TGT 0.38\n",
      "TGC 0.62\n",
      "TGA 0.91\n",
      "TGG 0.09\n",
      "CCT 0.22\n",
      "CCC 0.4\n",
      "CCA 0.37\n",
      "CCG 0.02\n",
      "CAT 0.34\n",
      "CAC 0.66\n",
      "CAA 0.95\n",
      "CAG 0.05\n",
      "CGT 0.17\n",
      "CGC 0.32\n",
      "CGA 0.45\n",
      "CGG 0.06\n",
      "ATT 0.46\n",
      "ATC 0.54\n",
      "ATA 0.85\n",
      "ATG 0.15\n",
      "ACT 0.21\n",
      "ACC 0.33\n",
      "ACA 0.44\n",
      "ACG 0.02\n",
      "AAT 0.33\n",
      "AAC 0.67\n",
      "AAA 0.93\n",
      "AAG 0.07\n",
      "GTT 0.23\n",
      "GTC 0.29\n",
      "GTA 0.4\n",
      "GTG 0.08\n",
      "GCT 0.26\n",
      "GCC 0.43\n",
      "GCA 0.3\n",
      "GCG 0.01\n",
      "GAT 0.45\n",
      "GAC 0.55\n",
      "GAA 0.79\n",
      "GAG 0.21\n",
      "GGT 0.22\n",
      "GGC 0.34\n",
      "GGA 0.32\n",
      "GGG 0.12\n",
      "TAA 0.85\n",
      "TAG 0.08\n",
      "AGA 0.0\n",
      "AGG 0.08\n"
     ]
    }
   ],
   "source": [
    "for key, value in fraction_dict.items():\n",
    "    print(key, value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls.append(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid = seq3(\"F\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Phe'"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amino_acid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
