{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Codon Usage and tRNA Anticodon\n",
    "\n",
    "Using a list of genbank files as input, this script:\n",
    "\n",
    "- [x] Extracts all its CDS's\n",
    "- [x] Calculates codon usage\n",
    "\n",
    "---\n",
    "\n",
    "- [x] Converts each sequence to fasta (or only the concatenated tRNAs)\n",
    "- [x] Uses the trnascan-SE to identify the anticodon\n",
    "\n",
    "---\n",
    "\n",
    "Generates an excel spreadsheet, containing the following fields:\n",
    "\n",
    "- [x] Species\n",
    "\n",
    "- [x] Translation table\n",
    "\n",
    "- [x] Aminoacid (Three-letter code)\n",
    "\n",
    "- [x] Codon for that aminoacid\n",
    "\n",
    "- [ ] Has anticodon in mito tRNAs? YES or EMPTY VALUE\n",
    "\n",
    "- [x] Number of occurrences of that codon in the genes \n",
    "\n",
    "- [x] \\1000 - (number_of_codon_occurences / total number of aa - including start_stop_codons) * 1000\n",
    "\n",
    "- [x] Fraction: codon occurences / sum of all codon occurences for that aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's import everything we need:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqUtils import seq3\n",
    "import subprocess, os\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order for this script to work, you have to execute it from within the directory with the genbank files or change the working directory using  **os.chdir(\"path/to/directory/\")**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to add all gb files **(no .gbk allowed!)** to a list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Allenopithecus_nigroviridis_NC_023965.1.gb',\n",
       " 'Allocebus_trichotis_NC_035631.1.gb',\n",
       " 'Allochrocebus_lhoesti_NC_023962.1.gb',\n",
       " 'Alouatta_caraya_NC_021938.1.gb',\n",
       " 'Alouatta_seniculus_NC_027825.1.gb',\n",
       " 'Aotus_azarai_NC_021939.1.gb',\n",
       " 'Aotus_azarai_azarai_NC_018115.1.gb',\n",
       " 'Aotus_lemurinus_NC_019799.1.gb',\n",
       " 'Aotus_nancymaae_NC_018116.1.gb',\n",
       " 'Ateles_belzebuth_NC_019800.1.gb',\n",
       " 'Avahi_laniger_NC_021940.1.gb',\n",
       " 'Cacajao_calvus_NC_021967.1.gb',\n",
       " 'Callimico_goeldii_NC_024628.1.gb',\n",
       " 'Callithrix_geoffroyi_NC_021941.1.gb',\n",
       " 'Callithrix_jacchus_NC_025586.1.gb',\n",
       " 'Callithrix_kuhlii_NC_027658.1.gb',\n",
       " 'Callithrix_penicillata_NC_030788.1.gb',\n",
       " 'Callithrix_pygmaea_NC_021942.1.gb',\n",
       " 'Carlito_syrichta_NC_012774.1.gb',\n",
       " 'Cebus_albifrons_NC_002763.1.gb',\n",
       " 'Cephalopachus_bancanus_NC_002811.1.gb',\n",
       " 'Cercocebus_atys_NC_028592.1.gb',\n",
       " 'Cercocebus_chrysogaster_NC_021943.1.gb',\n",
       " 'Cercocebus_torquatus_NC_023964.1.gb',\n",
       " 'Cercopithecus_albogularis_NC_021944.1.gb',\n",
       " 'Cercopithecus_diana_NC_023963.1.gb',\n",
       " 'Cercopithecus_mitis_NC_023961.1.gb',\n",
       " 'Cheirogaleus_crossleyi_NC_035604.1.gb',\n",
       " 'Cheirogaleus_major_NC_035655.1.gb',\n",
       " 'Cheirogaleus_medius_NC_021945.1.gb',\n",
       " 'Cheirogaleus_sibreei_NC_035605.1.gb',\n",
       " 'Cheracebus_lugens_NC_024630.1.gb',\n",
       " 'Chiropotes_albinasus_NC_021946.1.gb',\n",
       " 'Chiropotes_israelita_NC_024629.1.gb',\n",
       " 'Chlorocebus_aethiops_NC_007009.1.gb',\n",
       " 'Chlorocebus_aethiops_x_Chlorocebus_pygerythrus_NC_034276.1.gb',\n",
       " 'Chlorocebus_cynosuros_NC_024933.1.gb',\n",
       " 'Chlorocebus_djamdjamensis_NC_034277.1.gb',\n",
       " 'Chlorocebus_pygerythrus_NC_009747.1.gb',\n",
       " 'Chlorocebus_sabaeus_NC_008066.1.gb',\n",
       " 'Chlorocebus_tantalus_NC_009748.1.gb',\n",
       " 'Colobus_guereza_NC_006901.1.gb',\n",
       " 'Daubentonia_madagascariensis_NC_010299.1.gb',\n",
       " 'Erythrocebus_patas_NC_021947.1.gb',\n",
       " 'Eulemur_fulvus_fulvus_NC_012766.1.gb',\n",
       " 'Eulemur_fulvus_mayottensis_NC_012769.1.gb',\n",
       " 'Eulemur_macaco_macaco_NC_012771.1.gb',\n",
       " 'Eulemur_mongoz_NC_010300.1.gb',\n",
       " 'Eulemur_rubriventer_NC_026098.1.gb',\n",
       " 'Eulemur_rufus_NC_021948.1.gb',\n",
       " 'Galago_moholi_NC_021949.1.gb',\n",
       " 'Galago_senegalensis_NC_012761.1.gb',\n",
       " 'Gorilla_beringei_NC_037853.1.gb',\n",
       " 'Gorilla_gorilla_NC_001645.1.gb',\n",
       " 'Gorilla_gorilla_gorilla_NC_011120.1.gb',\n",
       " 'Hapalemur_griseus_NC_021950.1.gb',\n",
       " 'Homo_heidelbergensis_NC_023100.1.gb',\n",
       " 'Homo_sapiens_NC_012920.1.gb',\n",
       " 'Homo_sapiens_neanderthalensis_NC_011137.1.gb',\n",
       " 'Homo_sapiens_subsp._Denisova_NC_013993.1.gb',\n",
       " 'Hoolock_hoolock_NC_033885.1.gb',\n",
       " 'Hoolock_leuconedys_NC_033882.1.gb',\n",
       " 'Hoolock_leuconedys_x_Hoolock_tianxing_NC_033883.1.gb',\n",
       " 'Hoolock_tianxing_NC_033884.1.gb',\n",
       " 'Hylobates_agilis_NC_014042.1.gb',\n",
       " 'Hylobates_lar_NC_002082.1.gb',\n",
       " 'Hylobates_pileatus_NC_014045.1.gb',\n",
       " 'Indri_indri_NC_026095.1.gb',\n",
       " 'Lagothrix_lagotricha_NC_021951.1.gb',\n",
       " 'Lemur_catta_NC_004025.1.gb',\n",
       " 'Leontopithecus_chrysopygus_NC_037878.1.gb',\n",
       " 'Leontopithecus_rosalia_NC_021952.1.gb',\n",
       " 'Lepilemur_aeeclis_NC_034730.1.gb',\n",
       " 'Lepilemur_ahmansonorum_NC_034733.1.gb',\n",
       " 'Lepilemur_ankaranensis_NC_034717.1.gb',\n",
       " 'Lepilemur_betsileo_NC_034726.1.gb',\n",
       " 'Lepilemur_dorsalis_NC_034721.1.gb',\n",
       " 'Lepilemur_edwardsi_NC_034720.1.gb',\n",
       " 'Lepilemur_fleuretae_NC_034718.1.gb',\n",
       " 'Lepilemur_grewcockorum_NC_034729.1.gb',\n",
       " 'Lepilemur_hollandorum_NC_034738.1.gb',\n",
       " 'Lepilemur_hubbardorum_NC_014453.1.gb',\n",
       " 'Lepilemur_jamesorum_NC_034736.1.gb',\n",
       " 'Lepilemur_leucopus_NC_034719.1.gb',\n",
       " 'Lepilemur_microdon_NC_034739.1.gb',\n",
       " 'Lepilemur_milanoii_NC_034725.1.gb',\n",
       " 'Lepilemur_mittermeieri_NC_034728.1.gb',\n",
       " 'Lepilemur_otto_NC_034724.1.gb',\n",
       " 'Lepilemur_petteri_NC_034723.1.gb',\n",
       " 'Lepilemur_randrianasoloi_NC_034722.1.gb',\n",
       " 'Lepilemur_ruficaudatus_NC_021953.1.gb',\n",
       " 'Lepilemur_sahamalazensis_NC_034734.1.gb',\n",
       " 'Lepilemur_scottorum_NC_034737.1.gb',\n",
       " 'Lepilemur_seali_NC_034731.1.gb',\n",
       " 'Lepilemur_septentrionalis_NC_034727.1.gb',\n",
       " 'Lepilemur_tymerlachsoni_NC_034735.1.gb',\n",
       " 'Lepilemur_wrightae_NC_034732.1.gb',\n",
       " 'Lophocebus_aterrimus_NC_021954.1.gb',\n",
       " 'Loris_lydekkerianus_NC_021955.1.gb',\n",
       " 'Loris_tardigradus_NC_012763.1.gb',\n",
       " 'Macaca_arctoides_NC_025201.1.gb',\n",
       " 'Macaca_assamensis_NC_023795.1.gb',\n",
       " 'Macaca_cyclopis_NC_027449.1.gb',\n",
       " 'Macaca_fascicularis_NC_012670.1.gb',\n",
       " 'Macaca_fuscata_NC_025513.1.gb',\n",
       " 'Macaca_leonina_NC_027604.1.gb',\n",
       " 'Macaca_leucogenys_NC_031156.1.gb',\n",
       " 'Macaca_mulatta_NC_005943.1.gb',\n",
       " 'Macaca_mulatta_vestita_NC_037466.1.gb',\n",
       " 'Macaca_nemestrina_NC_026976.1.gb',\n",
       " 'Macaca_nigra_NC_026120.1.gb',\n",
       " 'Macaca_radiata_NC_036508.1.gb',\n",
       " 'Macaca_silenus_NC_025221.1.gb',\n",
       " 'Macaca_sinica_NC_037437.1.gb',\n",
       " 'Macaca_sylvanus_NC_002764.1.gb',\n",
       " 'Macaca_thibetana_NC_011519.1.gb',\n",
       " 'Macaca_tonkeana_NC_025222.1.gb',\n",
       " 'Mandrillus_leucophaeus_NC_028442.1.gb',\n",
       " 'Mandrillus_sphinx_NC_021956.1.gb',\n",
       " 'Megaladapis_edwardsi_NC_026088.1.gb',\n",
       " 'Microcebus_arnholdi_NC_035554.1.gb',\n",
       " 'Microcebus_berthae_NC_035629.1.gb',\n",
       " 'Microcebus_bongolavensis_NC_035559.1.gb',\n",
       " 'Microcebus_danfossi_NC_035600.1.gb',\n",
       " 'Microcebus_griseorufus_NC_035558.1.gb',\n",
       " 'Microcebus_jollyae_NC_035564.1.gb',\n",
       " 'Microcebus_lehilahytsara_NC_035602.1.gb',\n",
       " 'Microcebus_mamiratra_NC_035601.1.gb',\n",
       " 'Microcebus_margotmarshae_NC_035603.1.gb',\n",
       " 'Microcebus_mittermeieri_NC_035563.1.gb',\n",
       " 'Microcebus_murinus_NC_028718.1.gb',\n",
       " 'Microcebus_myoxinus_NC_035557.1.gb',\n",
       " 'Microcebus_ravelobensis_NC_035556.1.gb',\n",
       " 'Microcebus_rufus_NC_035562.1.gb',\n",
       " 'Microcebus_sambiranensis_NC_035561.1.gb',\n",
       " 'Microcebus_simmonsi_NC_035630.1.gb',\n",
       " 'Microcebus_tanosi_NC_035555.1.gb',\n",
       " 'Microcebus_tavaratra_NC_035560.1.gb',\n",
       " 'Mirza_coquereli_NC_035656.1.gb',\n",
       " 'Mirza_zaza_NC_035606.1.gb',\n",
       " 'Nasalis_larvatus_NC_008216.1.gb',\n",
       " 'Nomascus_gabriellae_NC_018753.1.gb',\n",
       " 'Nomascus_leucogenys_NC_021957.1.gb',\n",
       " 'Nomascus_siki_NC_014051.1.gb',\n",
       " 'Nycticebus_bengalensis_NC_021958.1.gb',\n",
       " 'Nycticebus_coucang_NC_002765.1.gb',\n",
       " 'Nycticebus_coucang_insularis_NC_040292.1.gb',\n",
       " 'Nycticebus_pygmaeus_NC_033381.1.gb',\n",
       " 'Otolemur_crassicaudatus_NC_012762.1.gb',\n",
       " 'Palaeopropithecus_ingens_NC_026090.1.gb',\n",
       " 'Pan_paniscus_NC_001644.1.gb',\n",
       " 'Pan_troglodytes_NC_001643.1.gb',\n",
       " 'Papio_anubis_NC_020006.2.gb',\n",
       " 'Papio_cynocephalus_NC_020007.2.gb',\n",
       " 'Papio_hamadryas_NC_001992.1.gb',\n",
       " 'Papio_kindae_NC_020008.2.gb',\n",
       " 'Papio_papio_NC_020009.2.gb',\n",
       " 'Papio_ursinus_NC_020010.2.gb',\n",
       " 'Perodicticus_potto_NC_012764.1.gb',\n",
       " 'Phaner_electromontis_NC_035607.1.gb',\n",
       " 'Phaner_pallescens_NC_035608.1.gb',\n",
       " 'Phaner_parienti_NC_035609.1.gb',\n",
       " 'Piliocolobus_badius_NC_008219.1.gb',\n",
       " 'Plecturocebus_cupreus_NC_021965.1.gb',\n",
       " 'Plecturocebus_donacophilus_NC_019801.1.gb',\n",
       " 'Pongo_abelii_NC_002083.1.gb',\n",
       " 'Pongo_pygmaeus_NC_001646.1.gb',\n",
       " 'Presbytis_melalophos_NC_008217.1.gb',\n",
       " 'Procolobus_verus_NC_020666.1.gb',\n",
       " 'Prolemur_simus_NC_021959.1.gb',\n",
       " 'Propithecus_coquereli_NC_011053.1.gb',\n",
       " 'Propithecus_diadema_NC_026084.1.gb',\n",
       " 'Propithecus_edwardsi_NC_026086.1.gb',\n",
       " 'Propithecus_tattersalli_NC_027740.1.gb',\n",
       " 'Propithecus_verreauxi_NC_028210.1.gb',\n",
       " 'Pygathrix_cinerea_1_RL-2012_NC_018062.1.gb',\n",
       " 'Pygathrix_cinerea_2_RL-2012_NC_018063.1.gb',\n",
       " 'Pygathrix_nemaeus_NC_008220.1.gb',\n",
       " 'Pygathrix_nigripes_NC_018061.1.gb',\n",
       " 'Rhinopithecus_avunculus_NC_015485.1.gb',\n",
       " 'Rhinopithecus_bieti_1_RL-2012_NC_018058.1.gb',\n",
       " 'Rhinopithecus_bieti_2_RL-2012_NC_018060.1.gb',\n",
       " 'Rhinopithecus_bieti_NC_015486.1.gb',\n",
       " 'Rhinopithecus_brelichi_NC_018057.1.gb',\n",
       " 'Rhinopithecus_roxellana_NC_008218.1.gb',\n",
       " 'Rhinopithecus_strykeri_NC_018059.1.gb',\n",
       " 'Saguinus_oedipus_NC_021960.1.gb',\n",
       " 'Saimiri_boliviensis_NC_021966.1.gb',\n",
       " 'Saimiri_boliviensis_boliviensis_NC_018096.1.gb',\n",
       " 'Saimiri_oerstedii_citrinellus_NC_023211.1.gb',\n",
       " 'Saimiri_sciureus_NC_012775.1.gb',\n",
       " 'Sapajus_xanthosternos_NC_021961.1.gb',\n",
       " 'Semnopithecus_entellus_NC_008215.1.gb',\n",
       " 'Simias_concolor_NC_020667.1.gb',\n",
       " 'Symphalangus_syndactylus_NC_014047.1.gb',\n",
       " 'Tarsius_dentatus_NC_024052.1.gb',\n",
       " 'Tarsius_lariang_NC_024051.1.gb',\n",
       " 'Tarsius_wallacei_NC_024053.1.gb',\n",
       " 'Theropithecus_gelada_NC_019802.1.gb',\n",
       " 'Trachypithecus_cristatus_NC_023971.1.gb',\n",
       " 'Trachypithecus_francoisi_NC_023970.1.gb',\n",
       " 'Trachypithecus_germaini_NC_019580.1.gb',\n",
       " 'Trachypithecus_hatinhensis_NC_019579.1.gb',\n",
       " 'Trachypithecus_johnii_NC_019583.1.gb',\n",
       " 'Trachypithecus_obscurus_NC_006900.1.gb',\n",
       " 'Trachypithecus_pileatus_NC_024529.1.gb',\n",
       " 'Trachypithecus_poliocephalus_NC_034795.1.gb',\n",
       " 'Trachypithecus_shortridgei_NC_019581.1.gb',\n",
       " 'Trachypithecus_vetulus_NC_019582.1.gb',\n",
       " 'Varecia_variegata_NC_026085.1.gb',\n",
       " 'Varecia_variegata_variegata_NC_012773.1.gb']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank = !ls *gb\n",
    "genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Allenopithecus_nigroviridis_NC_023965.1.gb'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# trnascan\n",
    "There are files for the different genetic tables in /home/gabriel/bioinfo/anaconda3/lib/tRNAscan-SE/gcode/gcode.vertmito.\n",
    "Maybe looking at this will give me some ideas on how to run the trnascan for other tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genbank_to_fasta(genbank):\n",
    "    record = SeqIO.read(genbank, \"genbank\")\n",
    "    species = record.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "    seq = record.seq\n",
    "    fasta = \">{}\\n{}\".format(species, seq)\n",
    "    return(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_trnascan(genbank):\n",
    "    fasta = genbank_to_fasta(genbank)\n",
    "    with open(\"tempfile\", \"w+\") as tempfile:\n",
    "        tempfile.write(fasta)\n",
    "        trnascan = subprocess.Popen([\"tRNAscan-SE\", \"-M\", \"mammal\", \"tempfile\"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "        trnascan.wait()\n",
    "        output, error = trnascan.communicate()\n",
    "        os.remove(\"tempfile\")\n",
    "        return(output.decode().split(\"\\n\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trnascan_se_anticodon_parser(trnascan_output, species):\n",
    "    anticodon_dict = dict()\n",
    "    for i in trnascan_output:\n",
    "         if i.startswith(species):\n",
    "                i = i.split(\"\\t\")\n",
    "                aminoacid = i[4].strip()\n",
    "                anticodon = i[5].strip()\n",
    "                if aminoacid not in anticodon_dict:\n",
    "                    anticodon_dict[aminoacid] = list()\n",
    "                anticodon_dict.get(aminoacid).append(anticodon)\n",
    "                \n",
    "    return(anticodon_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "We also have to calculate codon usage for all CDSs. In order to do so, we first need to extract all CDS from the genbank file, make sure that all sequences are multiple of 3 (or otherwise adding \"A\" residues from poly-A tail to complete the codon) and concatenate the CDS's."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_CDS(genbank):\n",
    "    CDS_list = list()\n",
    "    for record in SeqIO.parse(genbank, \"genbank\"):\n",
    "        accession = record.id\n",
    "        species_name = record.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for FEATURE in record.features:\n",
    "            if FEATURE.type == \"CDS\":\n",
    "                gene = FEATURE.qualifiers.get(\"gene\")[0]\n",
    "                sequence = FEATURE.location.extract(record).seq\n",
    "                CDS_list.append([accession, species_name, gene, sequence])\n",
    "    return(CDS_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_truncated(coding_list):\n",
    "    not_trunc_cds_list = list()\n",
    "    for i in coding_list:\n",
    "        truncated_nucs = len(i[3]) % 3\n",
    "        if truncated_nucs:\n",
    "            missing_nucs = 3 - truncated_nucs\n",
    "            #REWRITE THIS IN ORDER TO JUST ADD THE RESIDUES FROM THE POLY-A TAIL - DONE\n",
    "            #print(\"Truncated stop codon for gene {}\".format(i[2]))\n",
    "            #print(\"Before:{} truncated_nucs\".format(len(i[3]) % 3))\n",
    "            #print(\"Before: {}\".format(i[3]))\n",
    "            i[3] = i[3] + (missing_nucs * 'A')\n",
    "            #print()\n",
    "            #print(\"After: {}\".format(i[3]))\n",
    "            #print(\"After:{} truncated_nucs\".format(len(i[3]) % 3))\n",
    "            not_trunc_cds_list.append(i)\n",
    "        else:\n",
    "            not_trunc_cds_list.append(i)\n",
    "    return(not_trunc_cds_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NEXT STEP:\n",
    "\n",
    "Take a look at the \"CAI.py\" file and use it as a reference to write a function that counts codons for any genetic code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adapted from CAI.py:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from itertools import chain\n",
    "#from Bio.Data import CodonTable\n",
    "#from collections import Counter\n",
    "#from Bio.SeqUtils import seq3\n",
    "\n",
    "# get rid of Biopython warning\n",
    "#import warnings\n",
    "#from Bio import BiopythonWarning\n",
    "\n",
    "#warnings.simplefilter(\"ignore\", BiopythonWarning)\n",
    "\n",
    "\n",
    "def set_ncbi_genetic_codes():\n",
    "\n",
    "\n",
    "    ncbi_genetic_codes = {}\n",
    "    for table, codes in CodonTable.unambiguous_dna_by_id.items():\n",
    "        \n",
    "        # invert the genetic code dictionary to map each amino acid to its codons\n",
    "        # create dictionary of synonymous codons;\n",
    "        # Example: {'Phe': ['TTT', 'TTC'], ..., 'End': ['TAA', 'TAG', 'TGA']}\n",
    "        codons_for_amino_acid = {}\n",
    "        for codon, amino_acid in codes.forward_table.items():\n",
    "            amino_acid = seq3(amino_acid)\n",
    "            codons_for_amino_acid[amino_acid] = codons_for_amino_acid.get(amino_acid, [])\n",
    "            codons_for_amino_acid[amino_acid].append(codon)\n",
    "        #Adding the stop codons\n",
    "        codons_for_amino_acid[\"End\"] = CodonTable.unambiguous_dna_by_id[table].stop_codons\n",
    "        \n",
    "        #create dictionary of synonymous codons for each genetic table;\n",
    "        ncbi_genetic_codes[table] = codons_for_amino_acid\n",
    "        \n",
    "    return(ncbi_genetic_codes)\n",
    "\n",
    "#codon_tables = set_ncbi_genetic_codes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(codon_tables)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, all tables have exactly 64 codons, and 20 aminoacids (20 + stop) as expected:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'for table in codon_tables.keys():\\n    codons = 0\\n    for aa in codon_tables[table]:\\n        codons += len(codon_tables[table][aa])\\n    print(\"Table number: {} \\t Number of codons: {} \\t Aminoacids: {}\".format          (table, codons , len(codon_tables[table].keys())))'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''for table in codon_tables.keys():\n",
    "    codons = 0\n",
    "    for aa in codon_tables[table]:\n",
    "        codons += len(codon_tables[table][aa])\n",
    "    print(\"Table number: {} \\t Number of codons: {} \\t Aminoacids: {}\".format\\\n",
    "          (table, codons , len(codon_tables[table].keys())))'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_codons(sequence):\n",
    "    from itertools import product\n",
    "    \n",
    "    #create dictionary with each codon for counting (list comprehension inside dictionary comprehension)\n",
    "    codon_count_dict = {codon: 0 for codon in \n",
    "                  [''.join(i) for i in product(\"ATCG\", repeat=3)]}\n",
    "    \n",
    "    if len(sequence) % 3 != 0:\n",
    "        raise ValueError(\"Input sequence not divisible by three\")\n",
    "    for n, i in enumerate(range(0, len(sequence), 3), 1):\n",
    "        codon = str(sequence[i:i+3].upper())\n",
    "        if codon not in codon_count_dict.keys():\n",
    "            raise ValueError(\"This sequence probably has ambiguous IUPAC characters. \" \n",
    "                             \"Please submit sequences containing only A, T, G or C.\")\n",
    "        codon_count_dict[codon] += 1\n",
    "        \n",
    "    return(codon_count_dict)\n",
    "\n",
    "#print(count_codons(concat))\n",
    "\n",
    "#codon_count = count_codons(concat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NEXT STEP:\n",
    "Calculate all metrics associated with codon usage AND map it to its corresponding aminoacids, species, etc.\n",
    "Accomodate this data into a dataframe and plot it into an excel file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def per_thousand(codon_count_dict):\n",
    "    total_codon_number = sum(codon_count_dict.values())\n",
    "    per_thousand = {codon: round((count/total_codon_number)*1000, 2) \n",
    "                    for codon, count in codon_count_dict.items()}\n",
    "    return(per_thousand)    \n",
    "\n",
    "#per_thousand = per_thousand(codon_count)\n",
    "#print(per_thousand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_genetic_code(genbank):\n",
    "    record = SeqIO.read(genbank, \"genbank\")\n",
    "    genetic_code = set()\n",
    "    for i in record.features:\n",
    "        if i.type == \"CDS\":\n",
    "            genetic_code.add(i.qualifiers.get(\"transl_table\")[0])\n",
    "    if len(genetic_code) != 1:\n",
    "        raise ValueError(\"The {} file has more than one genetic code for its CDS's: {} \" \n",
    "                             \"Please correct that before submiting\".format(genbank, genetic_code))\n",
    "    return(int(list(genetic_code)[0]))\n",
    "    \n",
    "#genetic_code = identify_genetic_code(genbank[0]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(codon_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fraction(codon_count_dict, transl_table):\n",
    "    fraction = dict()\n",
    "    for codon_list in transl_table.values():\n",
    "        codon_occurrences_per_aa = 0\n",
    "        for codon in codon_list:\n",
    "            codon_occurrences_per_aa += codon_count_dict[codon]\n",
    "        for codon in codon_list:\n",
    "            fraction[codon] = round(codon_count_dict[codon]/codon_occurrences_per_aa, 2)\n",
    "    return(fraction)\n",
    "\n",
    "#fraction = fraction(codon_count, codon_tables[genetic_code])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trnascan_se_parser(trnascan_output, sequence_name):\n",
    "    aminoacid = list()\n",
    "    anticodon = list()\n",
    "    for i in trnascan_output:\n",
    "         if i.startswith(sequence_name):\n",
    "                i = i.split(\"\\t\")\n",
    "                aminoacid.append(i[4].strip())\n",
    "                anticodon.append(i[5].strip())\n",
    "    trnascan_dict = {sequence_name: {i[0]: i[1] for i in zip(aminoacid, anticodon)}}\n",
    "    return(trnascan_dict)\n",
    "        \n",
    "\n",
    "#trna_scan_parse = trnascan_se_parser(trnascan_output, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(trna_scan_parse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NEXT:\n",
    "Create main function and put all data into a dataframe.\n",
    "**TIP:** In the main function, put the codon usage functions first, then the anticodon.\n",
    "Structure chosen for the dataframe: list of dictionaries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'excel_dict = [\\n    {\\'species\\': species_name, \\'transl_table\\': translat_table,\\n     \\'aminoacid\\': Three-letter aa, \"Has anticodon in mito?\": Yes or no,\\n    \\'Number of codon occurences\\': number of codons, \\'/1000\\': per_thousand,\\n    \\'Fraction\\': fraction}\\n]'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#template:\n",
    "\n",
    "'''excel_dict = [\n",
    "    {'species': species_name, 'transl_table': translat_table,\n",
    "     'aminoacid': Three-letter aa, \"Has anticodon in mito?\": Yes or no,\n",
    "    'Number of codon occurences': number of codons, '/1000': per_thousand,\n",
    "    'Fraction': fraction}\n",
    "]'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [x] Species\n",
    "\n",
    "- [x] Translation table\n",
    "\n",
    "- [x] Aminoacid (Three-letter code)\n",
    "\n",
    "- [x] Codon for that aminoacid\n",
    "\n",
    "- [ ] Has anticodon in mito tRNAs? YES or EMPTY VALUE\n",
    "\n",
    "- [x] Number of occurrences of that codon in the genes \n",
    "\n",
    "- [x] \\1000 - (number_of_codon_occurences / total number of aa - including start_stop_codons) * 1000\n",
    "\n",
    "- [x] Fraction: codon occurences / sum of all codon occurences for that aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_CDS(not_trunc):\n",
    "    concat = ''\n",
    "    for i in not_trunc:\n",
    "        concat += i[3]\n",
    "    return concat\n",
    "\n",
    "def species_name(not_trunc):\n",
    "    species = set()\n",
    "    for i in not_trunc:\n",
    "        species.add(i[1])\n",
    "    if len(species) == 1:\n",
    "        return list(species)[0]\n",
    "    else:\n",
    "        raise ValueError(\"More than one organism for a single genbank file?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def invert_codon_table(codon_table):\n",
    "    new_codon_table = dict()\n",
    "    for key in codon_table.keys():\n",
    "        #print(key)\n",
    "        for codon in codon_table.get(key):\n",
    "            new_codon_table[codon] = key\n",
    "            \n",
    "    return(new_codon_table)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_anticodon(codon, aa_3letter, anticodon_dict):\n",
    "    #print(aa_3letter)\n",
    "    if aa_3letter in anticodon_dict.keys():\n",
    "        for value in anticodon_dict.get(aa_3letter):\n",
    "            anticodon = Seq(value, generic_dna)\n",
    "            if codon == str(anticodon.reverse_complement()):\n",
    "                return(value)\n",
    "    return(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dataframe_parser(species, genetic_code, inverse_codon_table, \n",
    "                     codon_count_dict, per_thousand_dict, fraction_dict, \n",
    "                     anticodon_dict):\n",
    "    fields = ['Species', 'Translation Table', 'Aminoacid', 'Codon', 'Anticodon', \n",
    "                  'Number of Codon Occurences', '/1000', 'Fraction']\n",
    "    dataframe_dict = {k : list() for k in fields}\n",
    "    for codon, fraction in fraction_dict.items():\n",
    "        dataframe_dict.get('Species').append(species)\n",
    "        dataframe_dict.get('Translation Table').append(genetic_code)\n",
    "        aa_3letter = inverse_codon_table.get(codon)\n",
    "        dataframe_dict.get('Aminoacid').append(aa_3letter)\n",
    "        dataframe_dict.get('Codon').append(codon)\n",
    "        dataframe_dict.get('Anticodon').append(has_anticodon(codon, aa_3letter, anticodon_dict))\n",
    "        dataframe_dict.get('Number of Codon Occurences').append(codon_count_dict.get(codon))\n",
    "        dataframe_dict.get('/1000').append(per_thousand_dict.get(codon))\n",
    "        dataframe_dict.get('Fraction').append(fraction)\n",
    "    return(dataframe_dict)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_dataframe(parsed_data):\n",
    "    codon_usage = pandas.DataFrame(parsed_data)\n",
    "    return(codon_usage)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_dataframes(dataframe_data):\n",
    "    dataframes = []\n",
    "    for i in dataframe_data.keys():\n",
    "        dataframes.append(dataframe_data.get(i))\n",
    "    final_dataframe = pandas.concat(dataframes)\n",
    "    return(final_dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_dataframe_to_excel(final_dataframe):\n",
    "    final_dataframe.to_excel(\"Codon_usage.xlsx\", index=False, sheet_name=\"codon_usage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running analysis for Allenopithecus_nigroviridis_NC_023965.1.gb\n",
      "Running analysis for Allocebus_trichotis_NC_035631.1.gb\n",
      "Running analysis for Allochrocebus_lhoesti_NC_023962.1.gb\n",
      "Running analysis for Alouatta_caraya_NC_021938.1.gb\n",
      "Running analysis for Alouatta_seniculus_NC_027825.1.gb\n",
      "Running analysis for Aotus_azarai_NC_021939.1.gb\n",
      "Running analysis for Aotus_azarai_azarai_NC_018115.1.gb\n",
      "Running analysis for Aotus_lemurinus_NC_019799.1.gb\n",
      "Running analysis for Aotus_nancymaae_NC_018116.1.gb\n",
      "Aotus_nancymaae_NC_018116.1.gb analysis failed\n",
      "Running analysis for Ateles_belzebuth_NC_019800.1.gb\n",
      "Running analysis for Avahi_laniger_NC_021940.1.gb\n",
      "Running analysis for Cacajao_calvus_NC_021967.1.gb\n",
      "Running analysis for Callimico_goeldii_NC_024628.1.gb\n",
      "Running analysis for Callithrix_geoffroyi_NC_021941.1.gb\n",
      "Running analysis for Callithrix_jacchus_NC_025586.1.gb\n",
      "Running analysis for Callithrix_kuhlii_NC_027658.1.gb\n",
      "Running analysis for Callithrix_penicillata_NC_030788.1.gb\n",
      "Running analysis for Callithrix_pygmaea_NC_021942.1.gb\n",
      "Running analysis for Carlito_syrichta_NC_012774.1.gb\n",
      "Running analysis for Cebus_albifrons_NC_002763.1.gb\n",
      "Running analysis for Cephalopachus_bancanus_NC_002811.1.gb\n",
      "Running analysis for Cercocebus_atys_NC_028592.1.gb\n",
      "Running analysis for Cercocebus_chrysogaster_NC_021943.1.gb\n",
      "Running analysis for Cercocebus_torquatus_NC_023964.1.gb\n",
      "Running analysis for Cercopithecus_albogularis_NC_021944.1.gb\n",
      "Running analysis for Cercopithecus_diana_NC_023963.1.gb\n",
      "Running analysis for Cercopithecus_mitis_NC_023961.1.gb\n",
      "Running analysis for Cheirogaleus_crossleyi_NC_035604.1.gb\n",
      "Running analysis for Cheirogaleus_major_NC_035655.1.gb\n",
      "Running analysis for Cheirogaleus_medius_NC_021945.1.gb\n",
      "Running analysis for Cheirogaleus_sibreei_NC_035605.1.gb\n",
      "Running analysis for Cheracebus_lugens_NC_024630.1.gb\n",
      "Running analysis for Chiropotes_albinasus_NC_021946.1.gb\n",
      "Running analysis for Chiropotes_israelita_NC_024629.1.gb\n",
      "Running analysis for Chlorocebus_aethiops_NC_007009.1.gb\n",
      "Chlorocebus_aethiops_NC_007009.1.gb analysis failed\n",
      "Running analysis for Chlorocebus_aethiops_x_Chlorocebus_pygerythrus_NC_034276.1.gb\n",
      "Running analysis for Chlorocebus_cynosuros_NC_024933.1.gb\n",
      "Running analysis for Chlorocebus_djamdjamensis_NC_034277.1.gb\n",
      "Running analysis for Chlorocebus_pygerythrus_NC_009747.1.gb\n",
      "Running analysis for Chlorocebus_sabaeus_NC_008066.1.gb\n",
      "Chlorocebus_sabaeus_NC_008066.1.gb analysis failed\n",
      "Running analysis for Chlorocebus_tantalus_NC_009748.1.gb\n",
      "Running analysis for Colobus_guereza_NC_006901.1.gb\n",
      "Running analysis for Daubentonia_madagascariensis_NC_010299.1.gb\n",
      "Running analysis for Erythrocebus_patas_NC_021947.1.gb\n",
      "Running analysis for Eulemur_fulvus_fulvus_NC_012766.1.gb\n",
      "Running analysis for Eulemur_fulvus_mayottensis_NC_012769.1.gb\n",
      "Running analysis for Eulemur_macaco_macaco_NC_012771.1.gb\n",
      "Running analysis for Eulemur_mongoz_NC_010300.1.gb\n",
      "Running analysis for Eulemur_rubriventer_NC_026098.1.gb\n",
      "Eulemur_rubriventer_NC_026098.1.gb analysis failed\n",
      "Running analysis for Eulemur_rufus_NC_021948.1.gb\n",
      "Running analysis for Galago_moholi_NC_021949.1.gb\n",
      "Running analysis for Galago_senegalensis_NC_012761.1.gb\n",
      "Running analysis for Gorilla_beringei_NC_037853.1.gb\n",
      "Running analysis for Gorilla_gorilla_NC_001645.1.gb\n",
      "Running analysis for Gorilla_gorilla_gorilla_NC_011120.1.gb\n",
      "Running analysis for Hapalemur_griseus_NC_021950.1.gb\n",
      "Running analysis for Homo_heidelbergensis_NC_023100.1.gb\n",
      "Homo_heidelbergensis_NC_023100.1.gb analysis failed\n",
      "Running analysis for Homo_sapiens_NC_012920.1.gb\n",
      "Running analysis for Homo_sapiens_neanderthalensis_NC_011137.1.gb\n",
      "Running analysis for Homo_sapiens_subsp._Denisova_NC_013993.1.gb\n",
      "Running analysis for Hoolock_hoolock_NC_033885.1.gb\n",
      "Hoolock_hoolock_NC_033885.1.gb analysis failed\n",
      "Running analysis for Hoolock_leuconedys_NC_033882.1.gb\n",
      "Running analysis for Hoolock_leuconedys_x_Hoolock_tianxing_NC_033883.1.gb\n",
      "Running analysis for Hoolock_tianxing_NC_033884.1.gb\n",
      "Running analysis for Hylobates_agilis_NC_014042.1.gb\n",
      "Running analysis for Hylobates_lar_NC_002082.1.gb\n",
      "Running analysis for Hylobates_pileatus_NC_014045.1.gb\n",
      "Running analysis for Indri_indri_NC_026095.1.gb\n",
      "Running analysis for Lagothrix_lagotricha_NC_021951.1.gb\n",
      "Running analysis for Lemur_catta_NC_004025.1.gb\n",
      "Running analysis for Leontopithecus_chrysopygus_NC_037878.1.gb\n",
      "Running analysis for Leontopithecus_rosalia_NC_021952.1.gb\n",
      "Running analysis for Lepilemur_aeeclis_NC_034730.1.gb\n",
      "Running analysis for Lepilemur_ahmansonorum_NC_034733.1.gb\n",
      "Running analysis for Lepilemur_ankaranensis_NC_034717.1.gb\n",
      "Running analysis for Lepilemur_betsileo_NC_034726.1.gb\n",
      "Running analysis for Lepilemur_dorsalis_NC_034721.1.gb\n",
      "Running analysis for Lepilemur_edwardsi_NC_034720.1.gb\n",
      "Running analysis for Lepilemur_fleuretae_NC_034718.1.gb\n",
      "Running analysis for Lepilemur_grewcockorum_NC_034729.1.gb\n",
      "Running analysis for Lepilemur_hollandorum_NC_034738.1.gb\n",
      "Running analysis for Lepilemur_hubbardorum_NC_014453.1.gb\n",
      "Running analysis for Lepilemur_jamesorum_NC_034736.1.gb\n",
      "Running analysis for Lepilemur_leucopus_NC_034719.1.gb\n",
      "Running analysis for Lepilemur_microdon_NC_034739.1.gb\n",
      "Running analysis for Lepilemur_milanoii_NC_034725.1.gb\n",
      "Running analysis for Lepilemur_mittermeieri_NC_034728.1.gb\n",
      "Running analysis for Lepilemur_otto_NC_034724.1.gb\n",
      "Running analysis for Lepilemur_petteri_NC_034723.1.gb\n",
      "Running analysis for Lepilemur_randrianasoloi_NC_034722.1.gb\n",
      "Running analysis for Lepilemur_ruficaudatus_NC_021953.1.gb\n",
      "Running analysis for Lepilemur_sahamalazensis_NC_034734.1.gb\n",
      "Running analysis for Lepilemur_scottorum_NC_034737.1.gb\n",
      "Running analysis for Lepilemur_seali_NC_034731.1.gb\n",
      "Running analysis for Lepilemur_septentrionalis_NC_034727.1.gb\n",
      "Running analysis for Lepilemur_tymerlachsoni_NC_034735.1.gb\n",
      "Running analysis for Lepilemur_wrightae_NC_034732.1.gb\n",
      "Running analysis for Lophocebus_aterrimus_NC_021954.1.gb\n",
      "Running analysis for Loris_lydekkerianus_NC_021955.1.gb\n",
      "Running analysis for Loris_tardigradus_NC_012763.1.gb\n",
      "Running analysis for Macaca_arctoides_NC_025201.1.gb\n",
      "Running analysis for Macaca_assamensis_NC_023795.1.gb\n",
      "Running analysis for Macaca_cyclopis_NC_027449.1.gb\n",
      "Macaca_cyclopis_NC_027449.1.gb analysis failed\n",
      "Running analysis for Macaca_fascicularis_NC_012670.1.gb\n",
      "Running analysis for Macaca_fuscata_NC_025513.1.gb\n",
      "Running analysis for Macaca_leonina_NC_027604.1.gb\n",
      "Running analysis for Macaca_leucogenys_NC_031156.1.gb\n",
      "Running analysis for Macaca_mulatta_NC_005943.1.gb\n",
      "Running analysis for Macaca_mulatta_vestita_NC_037466.1.gb\n",
      "Running analysis for Macaca_nemestrina_NC_026976.1.gb\n",
      "Running analysis for Macaca_nigra_NC_026120.1.gb\n",
      "Running analysis for Macaca_radiata_NC_036508.1.gb\n",
      "Running analysis for Macaca_silenus_NC_025221.1.gb\n",
      "Running analysis for Macaca_sinica_NC_037437.1.gb\n",
      "Running analysis for Macaca_sylvanus_NC_002764.1.gb\n",
      "Running analysis for Macaca_thibetana_NC_011519.1.gb\n",
      "Running analysis for Macaca_tonkeana_NC_025222.1.gb\n",
      "Running analysis for Mandrillus_leucophaeus_NC_028442.1.gb\n",
      "Running analysis for Mandrillus_sphinx_NC_021956.1.gb\n",
      "Running analysis for Megaladapis_edwardsi_NC_026088.1.gb\n",
      "Megaladapis_edwardsi_NC_026088.1.gb analysis failed\n",
      "Running analysis for Microcebus_arnholdi_NC_035554.1.gb\n",
      "Running analysis for Microcebus_berthae_NC_035629.1.gb\n",
      "Running analysis for Microcebus_bongolavensis_NC_035559.1.gb\n",
      "Running analysis for Microcebus_danfossi_NC_035600.1.gb\n",
      "Running analysis for Microcebus_griseorufus_NC_035558.1.gb\n",
      "Running analysis for Microcebus_jollyae_NC_035564.1.gb\n",
      "Running analysis for Microcebus_lehilahytsara_NC_035602.1.gb\n",
      "Running analysis for Microcebus_mamiratra_NC_035601.1.gb\n",
      "Running analysis for Microcebus_margotmarshae_NC_035603.1.gb\n",
      "Running analysis for Microcebus_mittermeieri_NC_035563.1.gb\n",
      "Running analysis for Microcebus_murinus_NC_028718.1.gb\n",
      "Running analysis for Microcebus_myoxinus_NC_035557.1.gb\n",
      "Running analysis for Microcebus_ravelobensis_NC_035556.1.gb\n",
      "Running analysis for Microcebus_rufus_NC_035562.1.gb\n",
      "Microcebus_rufus_NC_035562.1.gb analysis failed\n",
      "Running analysis for Microcebus_sambiranensis_NC_035561.1.gb\n",
      "Running analysis for Microcebus_simmonsi_NC_035630.1.gb\n",
      "Running analysis for Microcebus_tanosi_NC_035555.1.gb\n",
      "Running analysis for Microcebus_tavaratra_NC_035560.1.gb\n",
      "Running analysis for Mirza_coquereli_NC_035656.1.gb\n",
      "Running analysis for Mirza_zaza_NC_035606.1.gb\n",
      "Running analysis for Nasalis_larvatus_NC_008216.1.gb\n",
      "Running analysis for Nomascus_gabriellae_NC_018753.1.gb\n",
      "Nomascus_gabriellae_NC_018753.1.gb analysis failed\n",
      "Running analysis for Nomascus_leucogenys_NC_021957.1.gb\n",
      "Running analysis for Nomascus_siki_NC_014051.1.gb\n",
      "Running analysis for Nycticebus_bengalensis_NC_021958.1.gb\n",
      "Running analysis for Nycticebus_coucang_NC_002765.1.gb\n",
      "Running analysis for Nycticebus_coucang_insularis_NC_040292.1.gb\n",
      "Running analysis for Nycticebus_pygmaeus_NC_033381.1.gb\n",
      "Running analysis for Otolemur_crassicaudatus_NC_012762.1.gb\n",
      "Running analysis for Palaeopropithecus_ingens_NC_026090.1.gb\n",
      "Running analysis for Pan_paniscus_NC_001644.1.gb\n",
      "Running analysis for Pan_troglodytes_NC_001643.1.gb\n",
      "Running analysis for Papio_anubis_NC_020006.2.gb\n",
      "Running analysis for Papio_cynocephalus_NC_020007.2.gb\n",
      "Running analysis for Papio_hamadryas_NC_001992.1.gb\n",
      "Running analysis for Papio_kindae_NC_020008.2.gb\n",
      "Running analysis for Papio_papio_NC_020009.2.gb\n",
      "Running analysis for Papio_ursinus_NC_020010.2.gb\n",
      "Running analysis for Perodicticus_potto_NC_012764.1.gb\n",
      "Running analysis for Phaner_electromontis_NC_035607.1.gb\n",
      "Running analysis for Phaner_pallescens_NC_035608.1.gb\n",
      "Running analysis for Phaner_parienti_NC_035609.1.gb\n",
      "Running analysis for Piliocolobus_badius_NC_008219.1.gb\n",
      "Running analysis for Plecturocebus_cupreus_NC_021965.1.gb\n",
      "Running analysis for Plecturocebus_donacophilus_NC_019801.1.gb\n",
      "Running analysis for Pongo_abelii_NC_002083.1.gb\n",
      "Running analysis for Pongo_pygmaeus_NC_001646.1.gb\n",
      "Running analysis for Presbytis_melalophos_NC_008217.1.gb\n",
      "Running analysis for Procolobus_verus_NC_020666.1.gb\n",
      "Procolobus_verus_NC_020666.1.gb analysis failed\n",
      "Running analysis for Prolemur_simus_NC_021959.1.gb\n",
      "Running analysis for Propithecus_coquereli_NC_011053.1.gb\n",
      "Running analysis for Propithecus_diadema_NC_026084.1.gb\n",
      "Running analysis for Propithecus_edwardsi_NC_026086.1.gb\n",
      "Running analysis for Propithecus_tattersalli_NC_027740.1.gb\n",
      "Running analysis for Propithecus_verreauxi_NC_028210.1.gb\n",
      "Running analysis for Pygathrix_cinerea_1_RL-2012_NC_018062.1.gb\n",
      "Running analysis for Pygathrix_cinerea_2_RL-2012_NC_018063.1.gb\n",
      "Running analysis for Pygathrix_nemaeus_NC_008220.1.gb\n",
      "Running analysis for Pygathrix_nigripes_NC_018061.1.gb\n",
      "Running analysis for Rhinopithecus_avunculus_NC_015485.1.gb\n",
      "Running analysis for Rhinopithecus_bieti_1_RL-2012_NC_018058.1.gb\n",
      "Running analysis for Rhinopithecus_bieti_2_RL-2012_NC_018060.1.gb\n",
      "Running analysis for Rhinopithecus_bieti_NC_015486.1.gb\n",
      "Running analysis for Rhinopithecus_brelichi_NC_018057.1.gb\n",
      "Running analysis for Rhinopithecus_roxellana_NC_008218.1.gb\n",
      "Running analysis for Rhinopithecus_strykeri_NC_018059.1.gb\n",
      "Running analysis for Saguinus_oedipus_NC_021960.1.gb\n",
      "Running analysis for Saimiri_boliviensis_NC_021966.1.gb\n",
      "Running analysis for Saimiri_boliviensis_boliviensis_NC_018096.1.gb\n",
      "Running analysis for Saimiri_oerstedii_citrinellus_NC_023211.1.gb\n",
      "Running analysis for Saimiri_sciureus_NC_012775.1.gb\n",
      "Running analysis for Sapajus_xanthosternos_NC_021961.1.gb\n",
      "Running analysis for Semnopithecus_entellus_NC_008215.1.gb\n",
      "Running analysis for Simias_concolor_NC_020667.1.gb\n",
      "Running analysis for Symphalangus_syndactylus_NC_014047.1.gb\n",
      "Running analysis for Tarsius_dentatus_NC_024052.1.gb\n",
      "Running analysis for Tarsius_lariang_NC_024051.1.gb\n",
      "Running analysis for Tarsius_wallacei_NC_024053.1.gb\n",
      "Running analysis for Theropithecus_gelada_NC_019802.1.gb\n",
      "Running analysis for Trachypithecus_cristatus_NC_023971.1.gb\n",
      "Running analysis for Trachypithecus_francoisi_NC_023970.1.gb\n",
      "Running analysis for Trachypithecus_germaini_NC_019580.1.gb\n",
      "Running analysis for Trachypithecus_hatinhensis_NC_019579.1.gb\n",
      "Running analysis for Trachypithecus_johnii_NC_019583.1.gb\n",
      "Running analysis for Trachypithecus_obscurus_NC_006900.1.gb\n",
      "Running analysis for Trachypithecus_pileatus_NC_024529.1.gb\n",
      "Running analysis for Trachypithecus_poliocephalus_NC_034795.1.gb\n",
      "Running analysis for Trachypithecus_shortridgei_NC_019581.1.gb\n",
      "Running analysis for Trachypithecus_vetulus_NC_019582.1.gb\n",
      "Running analysis for Varecia_variegata_NC_026085.1.gb\n",
      "Varecia_variegata_NC_026085.1.gb analysis failed\n",
      "Running analysis for Varecia_variegata_variegata_NC_012773.1.gb\n",
      "                        Species  Translation Table Aminoacid Codon Anticodon  \\\n",
      "0   Allenopithecus_nigroviridis                  2       Phe   TTT      None   \n",
      "1   Allenopithecus_nigroviridis                  2       Phe   TTC       GAA   \n",
      "2   Allenopithecus_nigroviridis                  2       Leu   TTA       TAA   \n",
      "3   Allenopithecus_nigroviridis                  2       Leu   TTG      None   \n",
      "4   Allenopithecus_nigroviridis                  2       Leu   CTT      None   \n",
      "5   Allenopithecus_nigroviridis                  2       Leu   CTC      None   \n",
      "6   Allenopithecus_nigroviridis                  2       Leu   CTA       TAG   \n",
      "7   Allenopithecus_nigroviridis                  2       Leu   CTG      None   \n",
      "8   Allenopithecus_nigroviridis                  2       Ser   TCT      None   \n",
      "9   Allenopithecus_nigroviridis                  2       Ser   TCC      None   \n",
      "10  Allenopithecus_nigroviridis                  2       Ser   TCA       TGA   \n",
      "11  Allenopithecus_nigroviridis                  2       Ser   TCG      None   \n",
      "12  Allenopithecus_nigroviridis                  2       Ser   AGT      None   \n",
      "13  Allenopithecus_nigroviridis                  2       Ser   AGC       GCT   \n",
      "14  Allenopithecus_nigroviridis                  2       Tyr   TAT      None   \n",
      "15  Allenopithecus_nigroviridis                  2       Tyr   TAC       GTA   \n",
      "16  Allenopithecus_nigroviridis                  2       Cys   TGT      None   \n",
      "17  Allenopithecus_nigroviridis                  2       Cys   TGC       GCA   \n",
      "18  Allenopithecus_nigroviridis                  2       Trp   TGA       TCA   \n",
      "19  Allenopithecus_nigroviridis                  2       Trp   TGG      None   \n",
      "20  Allenopithecus_nigroviridis                  2       Pro   CCT      None   \n",
      "21  Allenopithecus_nigroviridis                  2       Pro   CCC      None   \n",
      "22  Allenopithecus_nigroviridis                  2       Pro   CCA       TGG   \n",
      "23  Allenopithecus_nigroviridis                  2       Pro   CCG      None   \n",
      "24  Allenopithecus_nigroviridis                  2       His   CAT      None   \n",
      "25  Allenopithecus_nigroviridis                  2       His   CAC       GTG   \n",
      "26  Allenopithecus_nigroviridis                  2       Gln   CAA       TTG   \n",
      "27  Allenopithecus_nigroviridis                  2       Gln   CAG      None   \n",
      "28  Allenopithecus_nigroviridis                  2       Arg   CGT      None   \n",
      "29  Allenopithecus_nigroviridis                  2       Arg   CGC      None   \n",
      "..                          ...                ...       ...   ...       ...   \n",
      "34  Varecia_variegata_variegata                  2       Met   ATA      None   \n",
      "35  Varecia_variegata_variegata                  2       Met   ATG       CAT   \n",
      "36  Varecia_variegata_variegata                  2       Thr   ACT      None   \n",
      "37  Varecia_variegata_variegata                  2       Thr   ACC      None   \n",
      "38  Varecia_variegata_variegata                  2       Thr   ACA       TGT   \n",
      "39  Varecia_variegata_variegata                  2       Thr   ACG      None   \n",
      "40  Varecia_variegata_variegata                  2       Asn   AAT      None   \n",
      "41  Varecia_variegata_variegata                  2       Asn   AAC       GTT   \n",
      "42  Varecia_variegata_variegata                  2       Lys   AAA       TTT   \n",
      "43  Varecia_variegata_variegata                  2       Lys   AAG      None   \n",
      "44  Varecia_variegata_variegata                  2       Val   GTT      None   \n",
      "45  Varecia_variegata_variegata                  2       Val   GTC      None   \n",
      "46  Varecia_variegata_variegata                  2       Val   GTA       TAC   \n",
      "47  Varecia_variegata_variegata                  2       Val   GTG      None   \n",
      "48  Varecia_variegata_variegata                  2       Ala   GCT      None   \n",
      "49  Varecia_variegata_variegata                  2       Ala   GCC      None   \n",
      "50  Varecia_variegata_variegata                  2       Ala   GCA       TGC   \n",
      "51  Varecia_variegata_variegata                  2       Ala   GCG      None   \n",
      "52  Varecia_variegata_variegata                  2       Asp   GAT      None   \n",
      "53  Varecia_variegata_variegata                  2       Asp   GAC       GTC   \n",
      "54  Varecia_variegata_variegata                  2       Glu   GAA       TTC   \n",
      "55  Varecia_variegata_variegata                  2       Glu   GAG      None   \n",
      "56  Varecia_variegata_variegata                  2       Gly   GGT      None   \n",
      "57  Varecia_variegata_variegata                  2       Gly   GGC      None   \n",
      "58  Varecia_variegata_variegata                  2       Gly   GGA       TCC   \n",
      "59  Varecia_variegata_variegata                  2       Gly   GGG      None   \n",
      "60  Varecia_variegata_variegata                  2       End   TAA      None   \n",
      "61  Varecia_variegata_variegata                  2       End   TAG      None   \n",
      "62  Varecia_variegata_variegata                  2       End   AGA      None   \n",
      "63  Varecia_variegata_variegata                  2       End   AGG      None   \n",
      "\n",
      "    Number of Codon Occurences  /1000  Fraction  \n",
      "0                           99  25.98      0.45  \n",
      "1                          120  31.50      0.55  \n",
      "2                          102  26.77      0.16  \n",
      "3                           17   4.46      0.03  \n",
      "4                           97  25.46      0.15  \n",
      "5                          103  27.03      0.16  \n",
      "6                          282  74.02      0.45  \n",
      "7                           26   6.82      0.04  \n",
      "8                           53  13.91      0.19  \n",
      "9                           81  21.26      0.29  \n",
      "10                          78  20.47      0.28  \n",
      "11                           9   2.36      0.03  \n",
      "12                          18   4.72      0.06  \n",
      "13                          39  10.24      0.14  \n",
      "14                          53  13.91      0.40  \n",
      "15                          81  21.26      0.60  \n",
      "16                           9   2.36      0.38  \n",
      "17                          15   3.94      0.62  \n",
      "18                          93  24.41      0.91  \n",
      "19                           9   2.36      0.09  \n",
      "20                          45  11.81      0.22  \n",
      "21                          82  21.52      0.40  \n",
      "22                          76  19.95      0.37  \n",
      "23                           4   1.05      0.02  \n",
      "24                          33   8.66      0.34  \n",
      "25                          65  17.06      0.66  \n",
      "26                          93  24.41      0.95  \n",
      "27                           5   1.31      0.05  \n",
      "28                          11   2.89      0.17  \n",
      "29                          21   5.51      0.32  \n",
      "..                         ...    ...       ...  \n",
      "34                         231  60.81      0.90  \n",
      "35                          26   6.84      0.10  \n",
      "36                          68  17.90      0.22  \n",
      "37                          97  25.53      0.32  \n",
      "38                         134  35.27      0.44  \n",
      "39                           7   1.84      0.02  \n",
      "40                          71  18.69      0.44  \n",
      "41                          92  24.22      0.56  \n",
      "42                          90  23.69      0.93  \n",
      "43                           7   1.84      0.07  \n",
      "44                          42  11.06      0.25  \n",
      "45                          33   8.69      0.20  \n",
      "46                          81  21.32      0.49  \n",
      "47                          11   2.90      0.07  \n",
      "48                          59  15.53      0.25  \n",
      "49                          94  24.74      0.40  \n",
      "50                          80  21.06      0.34  \n",
      "51                           1   0.26      0.00  \n",
      "52                          34   8.95      0.54  \n",
      "53                          29   7.63      0.46  \n",
      "54                          82  21.58      0.85  \n",
      "55                          14   3.69      0.15  \n",
      "56                          54  14.21      0.26  \n",
      "57                          55  14.48      0.26  \n",
      "58                          73  19.22      0.35  \n",
      "59                          29   7.63      0.14  \n",
      "60                           8   2.11      0.62  \n",
      "61                           3   0.79      0.23  \n",
      "62                           1   0.26      0.08  \n",
      "63                           1   0.26      0.08  \n",
      "\n",
      "[12736 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "def main_func(genbank_list):\n",
    "    codon_tables = set_ncbi_genetic_codes()\n",
    "    dataframe_data = dict()\n",
    "    for index, genbank in enumerate(genbank_list, 1):\n",
    "        print(\"Running analysis for {}\".format(genbank))\n",
    "        try:\n",
    "            not_trunc = fix_truncated(extract_CDS(genbank))\n",
    "        \n",
    "            cds_concat = concatenate_CDS(not_trunc)\n",
    "        \n",
    "            species = species_name(not_trunc)\n",
    "        \n",
    "            codon_count_dict = count_codons(cds_concat)\n",
    "        \n",
    "            per_thousand_dict = per_thousand(codon_count_dict)\n",
    "        \n",
    "            genetic_code = identify_genetic_code(genbank)\n",
    "                \n",
    "            fraction_dict = fraction(codon_count_dict, codon_tables[genetic_code])\n",
    "        \n",
    "            trnascan_output = run_trnascan(genbank)\n",
    "        \n",
    "            anticodon_dict = trnascan_se_anticodon_parser(trnascan_output, species)\n",
    "        \n",
    "            inverse_codon_table = invert_codon_table(codon_tables[genetic_code])\n",
    "        \n",
    "            parsed_data = dataframe_parser(species, genetic_code, inverse_codon_table, \n",
    "                                           codon_count_dict, per_thousand_dict, fraction_dict, \n",
    "                                           anticodon_dict)\n",
    "        except:\n",
    "            print(\"{} analysis failed\".format(genbank))\n",
    "            continue\n",
    "        \n",
    "        dataframe_data[index] = create_dataframe(parsed_data)\n",
    "        \n",
    "    final_dataframe = concat_dataframes(dataframe_data)\n",
    "    print(final_dataframe)\n",
    "    \n",
    "    export_dataframe_to_excel(final_dataframe)\n",
    "        \n",
    "        #print(anticodon_dict.keys())\n",
    "        #print(len(anticodon_dict.keys()))\n",
    "        #print(set(inverse_codon_table.values()))\n",
    "        #print(len(set(inverse_codon_table.values())))\n",
    "        #print(species, type(species))\n",
    "        #print(not_trunc)\n",
    "        #print(cds_concat)\n",
    "        #print(\"per_thousand_dict:\\n{}\".format(per_thousand_dict))\n",
    "        #print(\"codon_count_dict:\\n{}\".format(codon_count_dict))\n",
    "        #print(\"genetic_code {}:\\n{}\".format(genetic_code, codon_tables[genetic_code]))    \n",
    "        #print(\"fraction_dict:\\n{}\".format(fraction_dict))\n",
    "'''        fields = ['species', 'transl_table', 'aminoacid', \"Anticodon\", \n",
    "                  'Number of codon occurences', '/1000', 'Fraction']\n",
    "        excel_output_dict = {k : list() for k in fields}\n",
    "'''\n",
    "main_func(genbank)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Just me messing around with codon tables and such"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Phe': ['TTT', 'TTC'], 'Leu': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'Ser': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'], 'Tyr': ['TAT', 'TAC'], 'Cys': ['TGT', 'TGC'], 'Trp': ['TGA', 'TGG'], 'Pro': ['CCT', 'CCC', 'CCA', 'CCG'], 'His': ['CAT', 'CAC'], 'Gln': ['CAA', 'CAG'], 'Arg': ['CGT', 'CGC', 'CGA', 'CGG'], 'Ile': ['ATT', 'ATC'], 'Met': ['ATA', 'ATG'], 'Thr': ['ACT', 'ACC', 'ACA', 'ACG'], 'Asn': ['AAT', 'AAC'], 'Lys': ['AAA', 'AAG'], 'Val': ['GTT', 'GTC', 'GTA', 'GTG'], 'Ala': ['GCT', 'GCC', 'GCA', 'GCG'], 'Asp': ['GAT', 'GAC'], 'Glu': ['GAA', 'GAG'], 'Gly': ['GGT', 'GGC', 'GGA', 'GGG'], 'End': ['TAA', 'TAG', 'AGA', 'AGG']}\n"
     ]
    }
   ],
   "source": [
    "codon_tables = set_ncbi_genetic_codes()\n",
    "\n",
    "print(codon_tables[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Allochrocebus_lhoesti_NC_023962.1.gb\n"
     ]
    }
   ],
   "source": [
    "print(genbank[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "fields = ['species', 'transl_table', 'aminoacid', \"Anticodon\", \n",
    "                  'Number of codon occurences', '/1000', 'Fraction']\n",
    "excel_output_dict = {k : list() for k in fields}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTT 0.45\n",
      "TTC 0.55\n",
      "TTA 0.16\n",
      "TTG 0.03\n",
      "CTT 0.15\n",
      "CTC 0.16\n",
      "CTA 0.45\n",
      "CTG 0.04\n",
      "TCT 0.19\n",
      "TCC 0.29\n",
      "TCA 0.28\n",
      "TCG 0.03\n",
      "AGT 0.06\n",
      "AGC 0.14\n",
      "TAT 0.4\n",
      "TAC 0.6\n",
      "TGT 0.38\n",
      "TGC 0.62\n",
      "TGA 0.91\n",
      "TGG 0.09\n",
      "CCT 0.22\n",
      "CCC 0.4\n",
      "CCA 0.37\n",
      "CCG 0.02\n",
      "CAT 0.34\n",
      "CAC 0.66\n",
      "CAA 0.95\n",
      "CAG 0.05\n",
      "CGT 0.17\n",
      "CGC 0.32\n",
      "CGA 0.45\n",
      "CGG 0.06\n",
      "ATT 0.46\n",
      "ATC 0.54\n",
      "ATA 0.85\n",
      "ATG 0.15\n",
      "ACT 0.21\n",
      "ACC 0.33\n",
      "ACA 0.44\n",
      "ACG 0.02\n",
      "AAT 0.33\n",
      "AAC 0.67\n",
      "AAA 0.93\n",
      "AAG 0.07\n",
      "GTT 0.23\n",
      "GTC 0.29\n",
      "GTA 0.4\n",
      "GTG 0.08\n",
      "GCT 0.26\n",
      "GCC 0.43\n",
      "GCA 0.3\n",
      "GCG 0.01\n",
      "GAT 0.45\n",
      "GAC 0.55\n",
      "GAA 0.79\n",
      "GAG 0.21\n",
      "GGT 0.22\n",
      "GGC 0.34\n",
      "GGA 0.32\n",
      "GGG 0.12\n",
      "TAA 0.85\n",
      "TAG 0.08\n",
      "AGA 0.0\n",
      "AGG 0.08\n"
     ]
    }
   ],
   "source": [
    "for key, value in fraction_dict.items():\n",
    "    print(key, value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls.append(None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "amino_acid = seq3(\"F\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Phe'"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amino_acid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
